{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fd1eea88-aefd-46e6-8e95-1a6da6bdeef3",
   "metadata": {},
   "source": [
    "<div style=\"max-width:1200px\"><img src=\"assets/mgnify_banner.png\" width=\"100%\"></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ec2829d-7fb4-42c5-b27c-96327645128a",
   "metadata": {},
   "source": [
    "<img src=\"assets/mgnify_logo.png\" width=\"200px\">\n",
    "\n",
    "# Genome-resolved metagenomics bioinformatics 2022\n",
    "\n",
    "## Programmatic access to MGnify resources with APIs and notebooks (practical session)\n",
    "\n",
    "The [MGnify API](https://www.ebi.ac.uk/metagenomics/api/v1/) (Application Programming Interface) is a service to get data from MGnify using command line tools, scripts, code notebooks, etc.\n",
    "It can be access from pretty much any coding language... but Python and R are the most popular choices.\n",
    "\n",
    "### Today we will use Python\n",
    "<img src=\"https://www.python.org/static/community_logos/python-logo-master-v3-TM-flattened.png\" width=\"120px\">\n",
    "\n",
    "You don't need to know Python to do this — all of the answers are given.\n",
    "\n",
    "### Tomorrow we will use R\n",
    "<img src=\"https://www.r-project.org/logo/Rlogo.png\" width=\"60px\">\n",
    "\n",
    "You won't need to know R either – and there is a helper library called [MGnifyR](http://github.com/beadyallen/mgnifyr) that you'll learn to use tomorrow.\n",
    "\n",
    "**This is an interactive code notebook (a Jupyter Notebook). To run this code, click into each cell and press the ▶ button in the top toolbar, or press `shift + enter`**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fea10f7d-50d8-4e37-b123-a910b3782afd",
   "metadata": {},
   "source": [
    "---\n",
    "# Core concepts\n",
    "An [API](https://en.wikipedia.org/wiki/API \"Application programming interface\") is how your scripts (e.g. Python or R) can talk to the MGnify database.\n",
    "\n",
    "The MGnify API uses [JSON](https://en.wikipedia.org/wiki/JSON \"Javascript Object Notation\") to transfer data in a systematic way. This is human-readable and computer-readable.\n",
    "\n",
    "The particular format we use is a standard called [JSON:API](https://jsonapi.org). \n",
    "There is a Python package ([`jsonapi_client`](https://pypi.org/project/jsonapi-client/)) to make consuming this data easy. We're using it here.\n",
    "\n",
    "The MGnify API has a \"browsable interface\", which is a human-friendly way of exploring the API. The URLs for the browsable API are exactly the same as you'd use in a script or code; but when you open those URLs in a browser you see a nice interface. Find it here: [https://www.ebi.ac.uk/metagenomics/api/v1/](https://www.ebi.ac.uk/metagenomics/api/v1/).\n",
    "\n",
    "The MGnify API is \"paginated\", i.e. when you list some data you are given it in multiple pages. This is because there can sometimes by thousands of results. Thankfully `jsonapi_client` handles this for us."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f77f67e4-35cb-4253-bb95-4f2d0485c363",
   "metadata": {
    "tags": []
   },
   "source": [
    "---\n",
    "# Import packages"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e54652c-dcc6-4fc2-bb59-91ea66ae5730",
   "metadata": {
    "tags": []
   },
   "source": [
    "[pandas](https://pandas.pydata.org/docs/reference/index.html#api) is a data analysis library with a huge list of features. It is very good at holding and manipulating table data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5f8f45a-7bed-4f4b-a964-7813543e76de",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0199dc17-0761-4456-ae48-b9341a510e7f",
   "metadata": {},
   "source": [
    "[jsonapi-client](https://pypi.org/project/jsonapi-client/) is a library to get formatted data from web services into python code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "110553af-fab4-4159-9512-be796409d750",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from jsonapi_client import Session as APISession\n",
    "from jsonapi_client import Modifier"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3da4e633-63a6-4b4f-968a-583c172448b5",
   "metadata": {},
   "source": [
    "--- \n",
    "# API Discovery\n",
    "\n",
    "First, we will learn how to discover and use the API Endpoints (URLs) from the MGnify website.\n",
    "\n",
    "## When the API endpoint is directly mentioned\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "A task for you\n",
    "</div>\n",
    "\n",
    "- Go to [Study MGYS00002008](https://www.ebi.ac.uk/metagenomics/studies/MGYS00002008) on the MGnify website\n",
    "- Scroll down to the \"Programmatic access\" box\n",
    "- Find the \"API endpoint\" value\n",
    "- Click that API endpoint link, you'll see structured data – this is a JSON object\n",
    "- Back on the Study page, copy that API endpoint value\n",
    "- **See if you can complete the following code cell, which will read in that JSON data into Python...**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b573b529-30f9-4148-ae72-659027ac369c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "api_endpoint = 'https://_______________________________________'\n",
    "\n",
    "api_response = requests.get(api_endpoint)\n",
    "api_response.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7d1119b-c309-4bd4-89f0-5c8920472257",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "api_endpoint = 'https://_______________________________________'\n",
    "\n",
    "api_response = requests.get(api_endpoint)\n",
    "api_response.json()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "748b2e44-23d3-43fe-afb8-5a37cb9909a8",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "Unhide the cell below (click the •••) to see or use the answer.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5631662a-cec9-4796-9233-a18941db019e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'data': {'type': 'studies',\n",
       "  'id': 'MGYS00002008',\n",
       "  'attributes': {'bioproject': 'PRJEB22493',\n",
       "   'samples-count': 136,\n",
       "   'accession': 'MGYS00002008',\n",
       "   'is-private': False,\n",
       "   'secondary-accession': 'ERP104174',\n",
       "   'centre-name': 'EMBL-EBI',\n",
       "   'public-release-date': None,\n",
       "   'study-abstract': 'The APY Third Party Annotation (TPA) assembly was derived from the primary whole genome shotgun (WGS) data set PRJEB1787. This project includes samples from the following biomes : Marine.',\n",
       "   'study-name': 'EMG produced TPA metagenomics assembly of the Shotgun Sequencing of Tara Oceans DNA samples corresponding to size fractions for  prokaryotes. (APY) data set',\n",
       "   'data-origination': 'SUBMITTED',\n",
       "   'last-update': '2022-01-16T11:17:46'},\n",
       "  'relationships': {'studies': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00002008/studies'},\n",
       "    'data': [{'type': 'studies',\n",
       "      'id': 'MGYS00000410',\n",
       "      'links': {'self': 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00000410'}},\n",
       "     {'type': 'studies',\n",
       "      'id': 'MGYS00000492',\n",
       "      'links': {'self': 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00000492'}},\n",
       "     {'type': 'studies',\n",
       "      'id': 'MGYS00001482',\n",
       "      'links': {'self': 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00001482'}},\n",
       "     {'type': 'studies',\n",
       "      'id': 'MGYS00002392',\n",
       "      'links': {'self': 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00002392'}},\n",
       "     {'type': 'studies',\n",
       "      'id': 'MGYS00006075',\n",
       "      'links': {'self': 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00006075'}}]},\n",
       "   'samples': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00002008/samples'}},\n",
       "   'downloads': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00002008/downloads'}},\n",
       "   'biomes': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00002008/biomes'},\n",
       "    'data': [{'type': 'biomes',\n",
       "      'id': 'root:Environmental:Aquatic:Marine',\n",
       "      'links': {'self': 'https://www.ebi.ac.uk/metagenomics/api/v1/biomes/root:Environmental:Aquatic:Marine'}},\n",
       "     {'type': 'biomes',\n",
       "      'id': 'root:Environmental:Aquatic:Marine:Oceanic',\n",
       "      'links': {'self': 'https://www.ebi.ac.uk/metagenomics/api/v1/biomes/root:Environmental:Aquatic:Marine:Oceanic'}}]},\n",
       "   'analyses': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00002008/analyses'}},\n",
       "   'publications': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00002008/publications'}},\n",
       "   'geocoordinates': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00002008/geocoordinates'}}},\n",
       "  'links': {'self': 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00002008'}}}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "api_endpoint = 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00002008'\n",
    "\n",
    "api_response = requests.get(api_endpoint)\n",
    "api_response.json()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02ea0ff5-fd8b-4cb0-9bff-2c6f56f24968",
   "metadata": {},
   "source": [
    "Notice that this is now a Python dictionary of `data`, with some generic sounding top-level keys like `attributes` and `relationships`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6268fef6-df9a-4350-9bc8-4a2fb74b5fea",
   "metadata": {},
   "outputs": [],
   "source": [
    "api_response.json()['data'].keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "931bd460-b2b2-4dfa-8a8e-c9000d6f5bb2",
   "metadata": {},
   "source": [
    "These are part of the JSON:API specification that makes the API even easier to use with code libraries... More on that in a moment!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc8d002f-4cfe-4691-96dd-0ee6f41077dd",
   "metadata": {},
   "source": [
    "## From the Browsable API root\n",
    "Usually, you'll need to find a starting point for your API request endpoint.\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "A task for you\n",
    "</div>\n",
    "\n",
    "- Browse to https://www.ebi.ac.uk/metagenomics/api/v1\n",
    "- Notice that there are lots of API routes listed here, for all kind of MGnify datasets\n",
    "- Click on the `\"studies\"` endpoint\n",
    "- How would you find the MGYS00002008 endpoint from here?\n",
    "    - Click the \"Filters\" button near the top.\n",
    "    - Type `MGYS00002008` into the first filter (for an accession)\n",
    "    - Click Submit\n",
    "- Notice you're now on an API view which **lists** studies matching your query. There is only one, exactly matching your query, but it is still a list.\n",
    "- Find (by eye) the URL keyed by `data[0].links.self`.\n",
    "- Is this the same URL as the API Endpoint you copied from the website study page? (__It should be!__)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e59a5a4-b320-4a07-9e69-2bc0e5b23002",
   "metadata": {
    "tags": []
   },
   "source": [
    "## When you're after more than one database entry\n",
    "\n",
    "More realistically, you will run API queries that return many results.\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "A task for you\n",
    "</div>\n",
    "\n",
    "- Browse back to the [API Studies list endpoint](https://www.ebi.ac.uk/metagenomics/api/v1/studies)\n",
    "- Use the Filter button to limit the query to Studies from the `root:Host-associated:Insecta` biome lineage \n",
    "- The returned dataset is long... **but is it all of the matching Studies?**\n",
    "    - Check the `links` section of data... or the `meta.pagination` section of data (probably right at the bottom).\n",
    "    \n",
    "- To get **all** of the matching data, we need to make multiple requests - one for each page of data\n",
    "\n",
    "### Fetching paginated datasets\n",
    "This quickly gets complicated, but the `jsonapi_client` library can make it easy! It `iterate`s over the pages for us...\n",
    "\n",
    "**Try and complete this code cell to read all `root:Host-associated:Insecta` studies into a Pandas dataframe (like a table)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59c72da8-68a3-4819-b3d0-8ca6e671adac",
   "metadata": {},
   "outputs": [],
   "source": [
    "endpoint_name = \"studies\"\n",
    "\n",
    "with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    biome_filter = Modifier(\"lineage=root:______________________________\")\n",
    "    resources = map(lambda r: r.json, mgnify.iterate(endpoint_name, filter=biome_filter))\n",
    "    resources = pd.json_normalize(resources)\n",
    "resources"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8df0234a-8813-4074-9c30-6004691f1e79",
   "metadata": {},
   "source": [
    "If you know Pandas well... how would you write that table of studies into a CSV file?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b81e5b58-4cac-4d00-8dc7-65323a6673a8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# hint: there is a Pandas dataframe method to do exactly this!\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "879f3f44-d008-44cf-928c-810c59df3a90",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "Unhide the cell below (click the •••) to see or use the answer.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2603ae8e-5e38-4cbd-8145-3926737e86d1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.bioproject</th>\n",
       "      <th>attributes.accession</th>\n",
       "      <th>attributes.samples-count</th>\n",
       "      <th>attributes.is-private</th>\n",
       "      <th>attributes.secondary-accession</th>\n",
       "      <th>attributes.centre-name</th>\n",
       "      <th>attributes.study-abstract</th>\n",
       "      <th>attributes.study-name</th>\n",
       "      <th>attributes.data-origination</th>\n",
       "      <th>attributes.last-update</th>\n",
       "      <th>relationships.biomes.data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00006026</td>\n",
       "      <td>PRJEB53590</td>\n",
       "      <td>MGYS00006026</td>\n",
       "      <td>14</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP138399</td>\n",
       "      <td>EMG</td>\n",
       "      <td>The Third Party Annotation (TPA) assembly was ...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of PRJN...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2022-06-29T18:39:12</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta:Digestiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00006019</td>\n",
       "      <td>PRJNA422176</td>\n",
       "      <td>MGYS00006019</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>SRP126632</td>\n",
       "      <td>KAUST</td>\n",
       "      <td>The gut bacterial microbiome of honeybee (Apis...</td>\n",
       "      <td>Honeybee bacterial community across gut compar...</td>\n",
       "      <td>HARVESTED</td>\n",
       "      <td>2022-05-26T10:45:22</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta:Digestiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00006022</td>\n",
       "      <td>PRJNA422177</td>\n",
       "      <td>MGYS00006022</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>SRP126630</td>\n",
       "      <td>KAUST</td>\n",
       "      <td>The gut fungal microbiome of honeybee (Apis me...</td>\n",
       "      <td>Honeybee fungal community across gut compartments</td>\n",
       "      <td>HARVESTED</td>\n",
       "      <td>2022-05-26T10:36:25</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta:Digestiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00005862</td>\n",
       "      <td>PRJNA660574</td>\n",
       "      <td>MGYS00005862</td>\n",
       "      <td>21</td>\n",
       "      <td>False</td>\n",
       "      <td>SRP279534</td>\n",
       "      <td>IRTA-CReSA</td>\n",
       "      <td>The microbiome profile of a local population o...</td>\n",
       "      <td>Microbiota of field-caught Anopheles atroparvus</td>\n",
       "      <td>HARVESTED</td>\n",
       "      <td>2022-01-06T15:23:37</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00005463</td>\n",
       "      <td>PRJEB31429</td>\n",
       "      <td>MGYS00005463</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP113989</td>\n",
       "      <td>EMG</td>\n",
       "      <td>The Third Party Annotation (TPA) assembly was ...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2020-05-22T08:53:19</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00005464</td>\n",
       "      <td>PRJNA407112</td>\n",
       "      <td>MGYS00005464</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>SRP117625</td>\n",
       "      <td>Columbia</td>\n",
       "      <td>The ecology of bee viruses is complex, as viru...</td>\n",
       "      <td>Investigating the viral ecology of global bee ...</td>\n",
       "      <td>HARVESTED</td>\n",
       "      <td>2020-05-22T03:34:31</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00005215</td>\n",
       "      <td>PRJNA309422</td>\n",
       "      <td>MGYS00005215</td>\n",
       "      <td>424</td>\n",
       "      <td>False</td>\n",
       "      <td>SRP071118</td>\n",
       "      <td>University of Texas</td>\n",
       "      <td>16S rRNA amplicons from corbiculate bee gut co...</td>\n",
       "      <td>Corbiculate bee gut amplicons Raw sequence reads</td>\n",
       "      <td>HARVESTED</td>\n",
       "      <td>2020-05-02T01:47:47</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta:Digestiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00005163</td>\n",
       "      <td>PRJEB13694</td>\n",
       "      <td>MGYS00005163</td>\n",
       "      <td>429</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP015256</td>\n",
       "      <td>MAX PLANCK INSTITUTE FOR DEVELOPMENTAL BIOLOGY</td>\n",
       "      <td>Insects and nematodes represent the most speci...</td>\n",
       "      <td>Metagenomic analysis of nematode beetle associ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2020-02-10T10:53:57</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00001047</td>\n",
       "      <td>PRJNA266104</td>\n",
       "      <td>MGYS00001047</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>SRP050069</td>\n",
       "      <td>University of Innsbruck</td>\n",
       "      <td>Soil dwelling Dipteran larvae of Dilophus febr...</td>\n",
       "      <td>Microbiota of Dipteran gut, plant litter and c...</td>\n",
       "      <td>HARVESTED</td>\n",
       "      <td>2019-11-07T16:56:18</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta:Digestiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00005050</td>\n",
       "      <td>PRJEB27734</td>\n",
       "      <td>MGYS00005050</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP109856</td>\n",
       "      <td>Lawrence Berkeley National Laboratory</td>\n",
       "      <td>Metagenomic characterization of the 4 gut regi...</td>\n",
       "      <td>Passalid Beetle2</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2019-06-28T00:00:00</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta:Digestiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00003800</td>\n",
       "      <td>PRJEB21443</td>\n",
       "      <td>MGYS00003800</td>\n",
       "      <td>154</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP023686</td>\n",
       "      <td>UCSDMI</td>\n",
       "      <td>The sleep deprivation experiment was done in w...</td>\n",
       "      <td>Dominguez sleep deprived flies</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2019-02-19T18:44:34</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00003735</td>\n",
       "      <td>PRJEB29073</td>\n",
       "      <td>MGYS00003735</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP111342</td>\n",
       "      <td>KING ABDULAZIZ UNIVERSITY</td>\n",
       "      <td>Migratory locust gut microbiome isolated from ...</td>\n",
       "      <td>Migratory locust</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2019-02-18T11:31:16</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta:Digestiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00003705</td>\n",
       "      <td>PRJEB14927</td>\n",
       "      <td>MGYS00003705</td>\n",
       "      <td>388</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP016607</td>\n",
       "      <td>UCSDMI</td>\n",
       "      <td>Samples of bees, royal jelly from honey bees (...</td>\n",
       "      <td>Microbiome of honey bees from Puerto Rico</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-12-19T14:30:20</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00003201</td>\n",
       "      <td>PRJEB26091</td>\n",
       "      <td>MGYS00003201</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP108064</td>\n",
       "      <td>EMG</td>\n",
       "      <td>The Blattabacterium cuenoti BPAY Third Party A...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-10-25T15:57:57</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta:Digestiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00002713</td>\n",
       "      <td>PRJEB28193</td>\n",
       "      <td>MGYS00002713</td>\n",
       "      <td>255</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP110369</td>\n",
       "      <td>Uppsala University, ICM, Microbiology</td>\n",
       "      <td>The aim was to investigate the effect of genet...</td>\n",
       "      <td>Effect of genetics and environment on gut bact...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-08-29T14:55:19</td>\n",
       "      <td>[{'id': 'root:Engineered:Lab enrichment:Define...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00002336</td>\n",
       "      <td>PRJEB24507</td>\n",
       "      <td>MGYS00002336</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP106338</td>\n",
       "      <td>EMBL-EBI</td>\n",
       "      <td>The insect metagenome Third Party Annotation (...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-04-12T10:14:36</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00002333</td>\n",
       "      <td>PRJEB24111</td>\n",
       "      <td>MGYS00002333</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP105916</td>\n",
       "      <td>EMBL-EBI</td>\n",
       "      <td>The Bacteria Third Party Annotation (TPA) asse...</td>\n",
       "      <td>EMG produced TPA metagenomics assembly of the ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-04-12T07:47:35</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00002114</td>\n",
       "      <td>PRJEB24265</td>\n",
       "      <td>MGYS00002114</td>\n",
       "      <td>9</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP106076</td>\n",
       "      <td>University of Gdansk</td>\n",
       "      <td>microbiome profile of fungivorous thrips Hoplo...</td>\n",
       "      <td>microbiome profile of fungivorous thrips Hoplo...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2018-01-11T10:53:14</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta:Digestiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00002085</td>\n",
       "      <td>PRJEB23860</td>\n",
       "      <td>MGYS00002085</td>\n",
       "      <td>21</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP105639</td>\n",
       "      <td>University of Gdansk</td>\n",
       "      <td>In the present study we used Next Generation S...</td>\n",
       "      <td>First insight into microbial communities assoc...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2017-12-15T16:08:48</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00001998</td>\n",
       "      <td>PRJEB23277</td>\n",
       "      <td>MGYS00001998</td>\n",
       "      <td>129</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP105016</td>\n",
       "      <td>Institut de recherche experimentale et cliniqu...</td>\n",
       "      <td>Insects are frequently associated with bacteri...</td>\n",
       "      <td>A large-scale field study of bacterial communi...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2017-11-06T10:35:44</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta:Digestiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00001997</td>\n",
       "      <td>PRJEB22873</td>\n",
       "      <td>MGYS00001997</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP104589</td>\n",
       "      <td>University of Gdansk</td>\n",
       "      <td>In the present study we used Next Generation S...</td>\n",
       "      <td>First insight into microbiome profile of fungi...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2017-11-06T09:18:51</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00001961</td>\n",
       "      <td>PRJEB22282</td>\n",
       "      <td>MGYS00001961</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP103960</td>\n",
       "      <td>JOHN INNES CENTRE, NORWICH, UK.</td>\n",
       "      <td>Gut bacteria from giant stick insect guts</td>\n",
       "      <td>Gut bacteria from giant stick insect guts</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2017-10-04T13:47:53</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta:Digestiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00001896</td>\n",
       "      <td>PRJEB21763</td>\n",
       "      <td>MGYS00001896</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP024049</td>\n",
       "      <td>LSA-DISTAL-UNIBO</td>\n",
       "      <td>Tick</td>\n",
       "      <td>Tick</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2017-07-21T08:51:00</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00001785</td>\n",
       "      <td>PRJEB9958</td>\n",
       "      <td>MGYS00001785</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP011118</td>\n",
       "      <td>UNIVERSITY OF SOUTH BOHEMIA</td>\n",
       "      <td>Symbiosis between insects and bacteria result ...</td>\n",
       "      <td>Arsenophonus and Sodalis Symbionts in Louse Fl...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2017-06-05T13:18:06</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta:Digestiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00001768</td>\n",
       "      <td>PRJEB17080</td>\n",
       "      <td>MGYS00001768</td>\n",
       "      <td>82</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP018941</td>\n",
       "      <td>LABORATORIO NACIONAL DE CIENCIA E TECNOLOGIA D...</td>\n",
       "      <td>Due to their ability in digesting lignocellulo...</td>\n",
       "      <td>ECOLOGICAL AND MOLECULAR ASPECTS OF LIGNOCELLU...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2017-05-24T14:46:10</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00001731</td>\n",
       "      <td>PRJEB7708</td>\n",
       "      <td>MGYS00001731</td>\n",
       "      <td>30</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP008645</td>\n",
       "      <td>IMPERIAL COLLEGE</td>\n",
       "      <td>16S pyrosequencing of the microbiota in Anophe...</td>\n",
       "      <td>16S pyrosequencing of the microbiota in Anophe...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2017-05-04T10:58:26</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00001726</td>\n",
       "      <td>PRJEB7417</td>\n",
       "      <td>MGYS00001726</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP007145</td>\n",
       "      <td>CENTRE FOR ECOLOGY AND HYDROLOGY</td>\n",
       "      <td>TODO</td>\n",
       "      <td>Bombus terrestris gut microbiome</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2017-05-02T11:31:44</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00001304</td>\n",
       "      <td>PRJEB16059</td>\n",
       "      <td>MGYS00001304</td>\n",
       "      <td>18</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP017901</td>\n",
       "      <td>Centre for Ecology and Conservation</td>\n",
       "      <td>Resistance to a range of insecticides is confe...</td>\n",
       "      <td>Can insecticide resistance alter behaviour and...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2016-11-14T16:55:30</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta:Digestiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00001232</td>\n",
       "      <td>PRJEB3414</td>\n",
       "      <td>MGYS00001232</td>\n",
       "      <td>14</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP002663</td>\n",
       "      <td>UV</td>\n",
       "      <td>Bacterial symbionts of insects have many metab...</td>\n",
       "      <td>Ontogeny of the gut microbiota in the cockroac...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2016-09-27T09:44:49</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta:Digestiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00001106</td>\n",
       "      <td>PRJEB14361</td>\n",
       "      <td>MGYS00001106</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP015998</td>\n",
       "      <td>Universita degli studi di Milano</td>\n",
       "      <td>The feeding behavior of insects may influence ...</td>\n",
       "      <td>To evaluate, using as model the major pest of ...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2016-06-22T09:56:48</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta:Digestiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00000361</td>\n",
       "      <td>PRJEB4985</td>\n",
       "      <td>MGYS00000361</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP004327</td>\n",
       "      <td>ARO-NYRC</td>\n",
       "      <td>Facultative bacterial endosymbionts are common...</td>\n",
       "      <td>Bacterial community composition of banded-wing...</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2016-01-20T14:12:06</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>studies</td>\n",
       "      <td>MGYS00000494</td>\n",
       "      <td>PRJEB1670</td>\n",
       "      <td>MGYS00000494</td>\n",
       "      <td>1564</td>\n",
       "      <td>False</td>\n",
       "      <td>ERP002372</td>\n",
       "      <td>SC</td>\n",
       "      <td>For information on this project, contributing ...</td>\n",
       "      <td>Anopheles_Genome_Variation_Project</td>\n",
       "      <td>SUBMITTED</td>\n",
       "      <td>2016-01-20T14:12:06</td>\n",
       "      <td>[{'id': 'root:Host-associated:Insecta', 'type'...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       type            id attributes.bioproject attributes.accession  \\\n",
       "0   studies  MGYS00006026            PRJEB53590         MGYS00006026   \n",
       "1   studies  MGYS00006019           PRJNA422176         MGYS00006019   \n",
       "2   studies  MGYS00006022           PRJNA422177         MGYS00006022   \n",
       "3   studies  MGYS00005862           PRJNA660574         MGYS00005862   \n",
       "4   studies  MGYS00005463            PRJEB31429         MGYS00005463   \n",
       "5   studies  MGYS00005464           PRJNA407112         MGYS00005464   \n",
       "6   studies  MGYS00005215           PRJNA309422         MGYS00005215   \n",
       "7   studies  MGYS00005163            PRJEB13694         MGYS00005163   \n",
       "8   studies  MGYS00001047           PRJNA266104         MGYS00001047   \n",
       "9   studies  MGYS00005050            PRJEB27734         MGYS00005050   \n",
       "10  studies  MGYS00003800            PRJEB21443         MGYS00003800   \n",
       "11  studies  MGYS00003735            PRJEB29073         MGYS00003735   \n",
       "12  studies  MGYS00003705            PRJEB14927         MGYS00003705   \n",
       "13  studies  MGYS00003201            PRJEB26091         MGYS00003201   \n",
       "14  studies  MGYS00002713            PRJEB28193         MGYS00002713   \n",
       "15  studies  MGYS00002336            PRJEB24507         MGYS00002336   \n",
       "16  studies  MGYS00002333            PRJEB24111         MGYS00002333   \n",
       "17  studies  MGYS00002114            PRJEB24265         MGYS00002114   \n",
       "18  studies  MGYS00002085            PRJEB23860         MGYS00002085   \n",
       "19  studies  MGYS00001998            PRJEB23277         MGYS00001998   \n",
       "20  studies  MGYS00001997            PRJEB22873         MGYS00001997   \n",
       "21  studies  MGYS00001961            PRJEB22282         MGYS00001961   \n",
       "22  studies  MGYS00001896            PRJEB21763         MGYS00001896   \n",
       "23  studies  MGYS00001785             PRJEB9958         MGYS00001785   \n",
       "24  studies  MGYS00001768            PRJEB17080         MGYS00001768   \n",
       "25  studies  MGYS00001731             PRJEB7708         MGYS00001731   \n",
       "26  studies  MGYS00001726             PRJEB7417         MGYS00001726   \n",
       "27  studies  MGYS00001304            PRJEB16059         MGYS00001304   \n",
       "28  studies  MGYS00001232             PRJEB3414         MGYS00001232   \n",
       "29  studies  MGYS00001106            PRJEB14361         MGYS00001106   \n",
       "30  studies  MGYS00000361             PRJEB4985         MGYS00000361   \n",
       "31  studies  MGYS00000494             PRJEB1670         MGYS00000494   \n",
       "\n",
       "    attributes.samples-count  attributes.is-private  \\\n",
       "0                         14                  False   \n",
       "1                         20                  False   \n",
       "2                         20                  False   \n",
       "3                         21                  False   \n",
       "4                          4                  False   \n",
       "5                          3                  False   \n",
       "6                        424                  False   \n",
       "7                        429                  False   \n",
       "8                          6                  False   \n",
       "9                          4                  False   \n",
       "10                       154                  False   \n",
       "11                         1                  False   \n",
       "12                       388                  False   \n",
       "13                         1                  False   \n",
       "14                       255                  False   \n",
       "15                         1                  False   \n",
       "16                         2                  False   \n",
       "17                         9                  False   \n",
       "18                        21                  False   \n",
       "19                       129                  False   \n",
       "20                         2                  False   \n",
       "21                         2                  False   \n",
       "22                         1                  False   \n",
       "23                         1                  False   \n",
       "24                        82                  False   \n",
       "25                        30                  False   \n",
       "26                         1                  False   \n",
       "27                        18                  False   \n",
       "28                        14                  False   \n",
       "29                         1                  False   \n",
       "30                         3                  False   \n",
       "31                      1564                  False   \n",
       "\n",
       "   attributes.secondary-accession  \\\n",
       "0                       ERP138399   \n",
       "1                       SRP126632   \n",
       "2                       SRP126630   \n",
       "3                       SRP279534   \n",
       "4                       ERP113989   \n",
       "5                       SRP117625   \n",
       "6                       SRP071118   \n",
       "7                       ERP015256   \n",
       "8                       SRP050069   \n",
       "9                       ERP109856   \n",
       "10                      ERP023686   \n",
       "11                      ERP111342   \n",
       "12                      ERP016607   \n",
       "13                      ERP108064   \n",
       "14                      ERP110369   \n",
       "15                      ERP106338   \n",
       "16                      ERP105916   \n",
       "17                      ERP106076   \n",
       "18                      ERP105639   \n",
       "19                      ERP105016   \n",
       "20                      ERP104589   \n",
       "21                      ERP103960   \n",
       "22                      ERP024049   \n",
       "23                      ERP011118   \n",
       "24                      ERP018941   \n",
       "25                      ERP008645   \n",
       "26                      ERP007145   \n",
       "27                      ERP017901   \n",
       "28                      ERP002663   \n",
       "29                      ERP015998   \n",
       "30                      ERP004327   \n",
       "31                      ERP002372   \n",
       "\n",
       "                               attributes.centre-name  \\\n",
       "0                                                 EMG   \n",
       "1                                               KAUST   \n",
       "2                                               KAUST   \n",
       "3                                          IRTA-CReSA   \n",
       "4                                                 EMG   \n",
       "5                                            Columbia   \n",
       "6                                 University of Texas   \n",
       "7      MAX PLANCK INSTITUTE FOR DEVELOPMENTAL BIOLOGY   \n",
       "8                             University of Innsbruck   \n",
       "9               Lawrence Berkeley National Laboratory   \n",
       "10                                             UCSDMI   \n",
       "11                          KING ABDULAZIZ UNIVERSITY   \n",
       "12                                             UCSDMI   \n",
       "13                                                EMG   \n",
       "14              Uppsala University, ICM, Microbiology   \n",
       "15                                           EMBL-EBI   \n",
       "16                                           EMBL-EBI   \n",
       "17                               University of Gdansk   \n",
       "18                               University of Gdansk   \n",
       "19  Institut de recherche experimentale et cliniqu...   \n",
       "20                               University of Gdansk   \n",
       "21                    JOHN INNES CENTRE, NORWICH, UK.   \n",
       "22                                   LSA-DISTAL-UNIBO   \n",
       "23                        UNIVERSITY OF SOUTH BOHEMIA   \n",
       "24  LABORATORIO NACIONAL DE CIENCIA E TECNOLOGIA D...   \n",
       "25                                   IMPERIAL COLLEGE   \n",
       "26                   CENTRE FOR ECOLOGY AND HYDROLOGY   \n",
       "27                Centre for Ecology and Conservation   \n",
       "28                                                 UV   \n",
       "29                   Universita degli studi di Milano   \n",
       "30                                           ARO-NYRC   \n",
       "31                                                 SC   \n",
       "\n",
       "                            attributes.study-abstract  \\\n",
       "0   The Third Party Annotation (TPA) assembly was ...   \n",
       "1   The gut bacterial microbiome of honeybee (Apis...   \n",
       "2   The gut fungal microbiome of honeybee (Apis me...   \n",
       "3   The microbiome profile of a local population o...   \n",
       "4   The Third Party Annotation (TPA) assembly was ...   \n",
       "5   The ecology of bee viruses is complex, as viru...   \n",
       "6   16S rRNA amplicons from corbiculate bee gut co...   \n",
       "7   Insects and nematodes represent the most speci...   \n",
       "8   Soil dwelling Dipteran larvae of Dilophus febr...   \n",
       "9   Metagenomic characterization of the 4 gut regi...   \n",
       "10  The sleep deprivation experiment was done in w...   \n",
       "11  Migratory locust gut microbiome isolated from ...   \n",
       "12  Samples of bees, royal jelly from honey bees (...   \n",
       "13  The Blattabacterium cuenoti BPAY Third Party A...   \n",
       "14  The aim was to investigate the effect of genet...   \n",
       "15  The insect metagenome Third Party Annotation (...   \n",
       "16  The Bacteria Third Party Annotation (TPA) asse...   \n",
       "17  microbiome profile of fungivorous thrips Hoplo...   \n",
       "18  In the present study we used Next Generation S...   \n",
       "19  Insects are frequently associated with bacteri...   \n",
       "20  In the present study we used Next Generation S...   \n",
       "21          Gut bacteria from giant stick insect guts   \n",
       "22                                               Tick   \n",
       "23  Symbiosis between insects and bacteria result ...   \n",
       "24  Due to their ability in digesting lignocellulo...   \n",
       "25  16S pyrosequencing of the microbiota in Anophe...   \n",
       "26                                               TODO   \n",
       "27  Resistance to a range of insecticides is confe...   \n",
       "28  Bacterial symbionts of insects have many metab...   \n",
       "29  The feeding behavior of insects may influence ...   \n",
       "30  Facultative bacterial endosymbionts are common...   \n",
       "31  For information on this project, contributing ...   \n",
       "\n",
       "                                attributes.study-name  \\\n",
       "0   EMG produced TPA metagenomics assembly of PRJN...   \n",
       "1   Honeybee bacterial community across gut compar...   \n",
       "2   Honeybee fungal community across gut compartments   \n",
       "3     Microbiota of field-caught Anopheles atroparvus   \n",
       "4   EMG produced TPA metagenomics assembly of the ...   \n",
       "5   Investigating the viral ecology of global bee ...   \n",
       "6    Corbiculate bee gut amplicons Raw sequence reads   \n",
       "7   Metagenomic analysis of nematode beetle associ...   \n",
       "8   Microbiota of Dipteran gut, plant litter and c...   \n",
       "9                                    Passalid Beetle2   \n",
       "10                     Dominguez sleep deprived flies   \n",
       "11                                   Migratory locust   \n",
       "12          Microbiome of honey bees from Puerto Rico   \n",
       "13  EMG produced TPA metagenomics assembly of the ...   \n",
       "14  Effect of genetics and environment on gut bact...   \n",
       "15  EMG produced TPA metagenomics assembly of the ...   \n",
       "16  EMG produced TPA metagenomics assembly of the ...   \n",
       "17  microbiome profile of fungivorous thrips Hoplo...   \n",
       "18  First insight into microbial communities assoc...   \n",
       "19  A large-scale field study of bacterial communi...   \n",
       "20  First insight into microbiome profile of fungi...   \n",
       "21          Gut bacteria from giant stick insect guts   \n",
       "22                                               Tick   \n",
       "23  Arsenophonus and Sodalis Symbionts in Louse Fl...   \n",
       "24  ECOLOGICAL AND MOLECULAR ASPECTS OF LIGNOCELLU...   \n",
       "25  16S pyrosequencing of the microbiota in Anophe...   \n",
       "26                   Bombus terrestris gut microbiome   \n",
       "27  Can insecticide resistance alter behaviour and...   \n",
       "28  Ontogeny of the gut microbiota in the cockroac...   \n",
       "29  To evaluate, using as model the major pest of ...   \n",
       "30  Bacterial community composition of banded-wing...   \n",
       "31                 Anopheles_Genome_Variation_Project   \n",
       "\n",
       "   attributes.data-origination attributes.last-update  \\\n",
       "0                    SUBMITTED    2022-06-29T18:39:12   \n",
       "1                    HARVESTED    2022-05-26T10:45:22   \n",
       "2                    HARVESTED    2022-05-26T10:36:25   \n",
       "3                    HARVESTED    2022-01-06T15:23:37   \n",
       "4                    SUBMITTED    2020-05-22T08:53:19   \n",
       "5                    HARVESTED    2020-05-22T03:34:31   \n",
       "6                    HARVESTED    2020-05-02T01:47:47   \n",
       "7                    SUBMITTED    2020-02-10T10:53:57   \n",
       "8                    HARVESTED    2019-11-07T16:56:18   \n",
       "9                    SUBMITTED    2019-06-28T00:00:00   \n",
       "10                   SUBMITTED    2019-02-19T18:44:34   \n",
       "11                   SUBMITTED    2019-02-18T11:31:16   \n",
       "12                   SUBMITTED    2018-12-19T14:30:20   \n",
       "13                   SUBMITTED    2018-10-25T15:57:57   \n",
       "14                   SUBMITTED    2018-08-29T14:55:19   \n",
       "15                   SUBMITTED    2018-04-12T10:14:36   \n",
       "16                   SUBMITTED    2018-04-12T07:47:35   \n",
       "17                   SUBMITTED    2018-01-11T10:53:14   \n",
       "18                   SUBMITTED    2017-12-15T16:08:48   \n",
       "19                   SUBMITTED    2017-11-06T10:35:44   \n",
       "20                   SUBMITTED    2017-11-06T09:18:51   \n",
       "21                   SUBMITTED    2017-10-04T13:47:53   \n",
       "22                   SUBMITTED    2017-07-21T08:51:00   \n",
       "23                   SUBMITTED    2017-06-05T13:18:06   \n",
       "24                   SUBMITTED    2017-05-24T14:46:10   \n",
       "25                   SUBMITTED    2017-05-04T10:58:26   \n",
       "26                   SUBMITTED    2017-05-02T11:31:44   \n",
       "27                   SUBMITTED    2016-11-14T16:55:30   \n",
       "28                   SUBMITTED    2016-09-27T09:44:49   \n",
       "29                   SUBMITTED    2016-06-22T09:56:48   \n",
       "30                   SUBMITTED    2016-01-20T14:12:06   \n",
       "31                   SUBMITTED    2016-01-20T14:12:06   \n",
       "\n",
       "                            relationships.biomes.data  \n",
       "0   [{'id': 'root:Host-associated:Insecta:Digestiv...  \n",
       "1   [{'id': 'root:Host-associated:Insecta:Digestiv...  \n",
       "2   [{'id': 'root:Host-associated:Insecta:Digestiv...  \n",
       "3   [{'id': 'root:Host-associated:Insecta', 'type'...  \n",
       "4   [{'id': 'root:Host-associated:Insecta', 'type'...  \n",
       "5   [{'id': 'root:Host-associated:Insecta', 'type'...  \n",
       "6   [{'id': 'root:Host-associated:Insecta:Digestiv...  \n",
       "7   [{'id': 'root:Host-associated:Insecta', 'type'...  \n",
       "8   [{'id': 'root:Host-associated:Insecta:Digestiv...  \n",
       "9   [{'id': 'root:Host-associated:Insecta:Digestiv...  \n",
       "10  [{'id': 'root:Host-associated:Insecta', 'type'...  \n",
       "11  [{'id': 'root:Host-associated:Insecta:Digestiv...  \n",
       "12  [{'id': 'root:Host-associated:Insecta', 'type'...  \n",
       "13  [{'id': 'root:Host-associated:Insecta:Digestiv...  \n",
       "14  [{'id': 'root:Engineered:Lab enrichment:Define...  \n",
       "15  [{'id': 'root:Host-associated:Insecta', 'type'...  \n",
       "16  [{'id': 'root:Host-associated:Insecta', 'type'...  \n",
       "17  [{'id': 'root:Host-associated:Insecta:Digestiv...  \n",
       "18  [{'id': 'root:Host-associated:Insecta', 'type'...  \n",
       "19  [{'id': 'root:Host-associated:Insecta:Digestiv...  \n",
       "20  [{'id': 'root:Host-associated:Insecta', 'type'...  \n",
       "21  [{'id': 'root:Host-associated:Insecta:Digestiv...  \n",
       "22  [{'id': 'root:Host-associated:Insecta', 'type'...  \n",
       "23  [{'id': 'root:Host-associated:Insecta:Digestiv...  \n",
       "24  [{'id': 'root:Host-associated:Insecta', 'type'...  \n",
       "25  [{'id': 'root:Host-associated:Insecta', 'type'...  \n",
       "26  [{'id': 'root:Host-associated:Insecta', 'type'...  \n",
       "27  [{'id': 'root:Host-associated:Insecta:Digestiv...  \n",
       "28  [{'id': 'root:Host-associated:Insecta:Digestiv...  \n",
       "29  [{'id': 'root:Host-associated:Insecta:Digestiv...  \n",
       "30  [{'id': 'root:Host-associated:Insecta', 'type'...  \n",
       "31  [{'id': 'root:Host-associated:Insecta', 'type'...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "endpoint_name = \"studies\"\n",
    "\n",
    "with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    biome_filter = Modifier(\"lineage=root:Host-associated:Insecta\")\n",
    "    resources = map(lambda r: r.json, mgnify.iterate(endpoint_name, filter=biome_filter))\n",
    "    resources = pd.json_normalize(resources)\n",
    "resources.to_csv(f\"insecta_studies.csv\")\n",
    "resources"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3d42d24-b5b0-43e9-93ca-352fb75e0ac5",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Line by line explanation\n",
    "\n",
    "```python\n",
    "### The packages were already imported, but if you wanted to use this snippet on it's own as a script you would import them like this:\n",
    "from jsonapi_client import Session as APISession\n",
    "import pandas as pd\n",
    "###\n",
    "\n",
    "\n",
    "endpoint_name = 'studies'\n",
    "# An \"endpoint\" is the specific resource within the API which we want to get data from. \n",
    "# It is the a URL relative to the \"server base URL\" of the API, which for MGnify is https://www.ebi.ac.uk/metagenomics/api/v1.\n",
    "# You can find the endpoints in the API Docs https://www.ebi.ac.uk/metagenomics/api/docs/ \n",
    "\n",
    "with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    # Calling \"APISession\" is enabling a connection to the MGnify API, that can be used multiple times. \n",
    "    # The `with...as mgnify` syntax is a Python \"context\". \n",
    "    # Everything inside the `with...` block (i.e. indented below it) can use the `APISession` which we've called `mgnify` here. \n",
    "    # When the `with` block closes (the indentation stops), the connection to the API is nicely cleaned up for us.\n",
    "    \n",
    "    # Using a Modifier, we can filter the results from the API. \n",
    "    # The biome_filter will add the \"lineage=XXX\" to the query sent to the API\n",
    "    # This will be used by the API to filter the studies in the response by the biome specified in \"lineage\"\n",
    "    biome_filter = Modifier(\"lineage=root:Host-associated:Insecta\")\n",
    "\n",
    "    resources = map(lambda r: r.json, mgnify.iterate(endpoint_name, filter=biome_filter))\n",
    "    # `map` applies a function to every element of an iterable - so do something to each thing in a list.\n",
    "    # Remember we said the API is paginated? \n",
    "    # `mgnify.iterate(endpoint_name)` is a very helpful function that loops over as many pages of results as there are.\n",
    "    # `lambda r: r.json` is grabbing the JSON attribute from each Super Study returned from the API.\n",
    "    # All together, this makes `resources` be a bunch of JSON representations we could loop through, each containing the data of a Super Study.\n",
    "    \n",
    "    resources = pd.json_normalize(resources)\n",
    "    # `pd` is the de-facto shorthand for the `pandas` package - you'll see it anywhere people are using pandas.\n",
    "    # The `json_normalize` function takes \"nested\" data and does its best to turn it into a table.\n",
    "    # You can throw quite strange-looking data at it and it usually does something sensible.\n",
    "    \n",
    "    resources.to_csv(f\"{endpoint_name}.csv\")\n",
    "    # Pandas has a built-in way of writing CSV (or TSV, etc) files, which is helpful for getting data into other tools.\n",
    "    # This writes the table-ified Super Study list to a file called `super-studies.csv`.\n",
    "    \n",
    "resources\n",
    "# In a Jupyter notebook, you can just write a variable name in a cell (or the last line of a long cell), and it will print it.\n",
    "# Jupyter knows how to display Pandas tables (actually called \"DataFrames\", because they are More Than Just Tables ™) in a pretty way.\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2df4b49-7bf0-4869-bf1b-05fd64e075ca",
   "metadata": {},
   "source": [
    "--- \n",
    "# MAGs resources - programmatically\n",
    "\n",
    "## Listing genomic resources from the MGnify API\n",
    "\n",
    "**The goals of this task are to make a genus-specific version of [Extended Data Figure 4a from Almeida et al 2020](https://www.nature.com/articles/s41587-020-0603-3/figures/10), as well as to see which biomes that genus is present in**\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "A task for you\n",
    "</div>\n",
    "\n",
    "- Write some Python code that fetches all MGnify genomes from the `Streptococcus` genus into a Pandas Dataframe.\n",
    "- Make a \"Box and whiskers\" plot of the Completeness and Contamination distributions for the genomes in that genus.\n",
    "- Make a histogram plot showing how many times this genus appears in each biome-specifc catalogue\n",
    "\n",
    "Hints:\n",
    "- The lineage for the Streptococcus genus is `d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus`\n",
    "- The [API endpoint name for genomes is `genomes`](https://www.ebi.ac.uk/metagenomics/api/v1/genomes)\n",
    "- The API filter key on that endpoint, for taxonomic lineage, is `taxon_lineage=`\n",
    "- The completeness and contamination % values for each genome are in `attributes['completeness']` and `attributes['contamination']` respectively. Pandas normalizes these to columns named `'attributes.completeness'` and `'attributes.contamination'`\n",
    "- There is a [Pandas method `.boxplot`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.boxplot.html) to make a boxplot from a dataframe column/series\n",
    "- The catalogue ID a genome belongs to is in `relationships['catalogue']['data']['id']`, which Pandas normalizes to a column named `relationships.catalogue.data.id`\n",
    "- A column (series) of data from a Pandas dataframe can be selected like `resources['my.column.name']` and the [Pandas `.hist()` method] by default shows a histogram of data counts grouped by unique values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7552b9fb-14fa-479d-acd7-c28eeb0e1b6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "endpoint_name = \"_______\"\n",
    "\n",
    "with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    lineage_filter = _________________________\n",
    "    resources = _________________\n",
    "    resources = pd.json_normalize(resources)\n",
    "resources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e18068bb-a9bd-4293-99fc-918d979875f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "_________.boxplot(_________________________)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01ed554a-db1c-4f55-b2df-f5ffdd0aaa25",
   "metadata": {},
   "outputs": [],
   "source": [
    "resources['____________________________').hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6be1c658-12a2-40c9-a7cf-876ced5f5960",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-info\">\n",
    "Unhide the cells below (click the •••) to see or use the answer.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "84be3663-9ba4-48ac-bae0-d750dd0c1ee7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.genome-id</th>\n",
       "      <th>attributes.geographic-range</th>\n",
       "      <th>attributes.geographic-origin</th>\n",
       "      <th>attributes.accession</th>\n",
       "      <th>attributes.ena-sample-accession</th>\n",
       "      <th>attributes.ena-study-accession</th>\n",
       "      <th>attributes.length</th>\n",
       "      <th>attributes.num-contigs</th>\n",
       "      <th>...</th>\n",
       "      <th>attributes.pangenome-accessory-size</th>\n",
       "      <th>attributes.last-update</th>\n",
       "      <th>attributes.first-created</th>\n",
       "      <th>relationships.catalogue.data.id</th>\n",
       "      <th>relationships.catalogue.data.type</th>\n",
       "      <th>relationships.biome.data.id</th>\n",
       "      <th>relationships.biome.data.type</th>\n",
       "      <th>attributes.ncbi-genome-accession</th>\n",
       "      <th>attributes.ncbi-sample-accession</th>\n",
       "      <th>attributes.ncbi-study-accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG000301840</td>\n",
       "      <td>14521</td>\n",
       "      <td>[Asia]</td>\n",
       "      <td>Asia</td>\n",
       "      <td>MGYG000301840</td>\n",
       "      <td>ERS12510205</td>\n",
       "      <td>ERP138410</td>\n",
       "      <td>2399126</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>1072.0</td>\n",
       "      <td>2022-11-29T11:09:03.681599</td>\n",
       "      <td>2022-11-29T11:09:03.678953</td>\n",
       "      <td>pig-gut-v1-0</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>root:Host-associated:Mammals:Gastrointestinal ...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG000301652</td>\n",
       "      <td>15565</td>\n",
       "      <td>[Asia]</td>\n",
       "      <td>Asia</td>\n",
       "      <td>MGYG000301652</td>\n",
       "      <td>ERS12501501</td>\n",
       "      <td>ERP137783</td>\n",
       "      <td>1861661</td>\n",
       "      <td>46</td>\n",
       "      <td>...</td>\n",
       "      <td>206.0</td>\n",
       "      <td>2022-11-29T11:38:29.625963</td>\n",
       "      <td>2022-11-29T11:38:29.623194</td>\n",
       "      <td>pig-gut-v1-0</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>root:Host-associated:Mammals:Gastrointestinal ...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG000301379</td>\n",
       "      <td>15032</td>\n",
       "      <td>[Europe]</td>\n",
       "      <td>Europe</td>\n",
       "      <td>MGYG000301379</td>\n",
       "      <td>ERS12500223</td>\n",
       "      <td>ERP137406</td>\n",
       "      <td>1772300</td>\n",
       "      <td>209</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2022-11-29T11:23:07.346774</td>\n",
       "      <td>2022-11-29T11:23:07.342356</td>\n",
       "      <td>pig-gut-v1-0</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>root:Host-associated:Mammals:Gastrointestinal ...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG000301371</td>\n",
       "      <td>14811</td>\n",
       "      <td>[Europe, Asia]</td>\n",
       "      <td>Europe</td>\n",
       "      <td>MGYG000301371</td>\n",
       "      <td>ERS12500425</td>\n",
       "      <td>ERP137406</td>\n",
       "      <td>1211138</td>\n",
       "      <td>255</td>\n",
       "      <td>...</td>\n",
       "      <td>627.0</td>\n",
       "      <td>2022-11-29T11:17:01.363252</td>\n",
       "      <td>2022-11-29T11:17:01.360465</td>\n",
       "      <td>pig-gut-v1-0</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>root:Host-associated:Mammals:Gastrointestinal ...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG000301270</td>\n",
       "      <td>15546</td>\n",
       "      <td>[]</td>\n",
       "      <td>Asia</td>\n",
       "      <td>MGYG000301270</td>\n",
       "      <td>ERS12500731</td>\n",
       "      <td>ERP137371</td>\n",
       "      <td>1887219</td>\n",
       "      <td>28</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2022-11-29T11:37:56.513553</td>\n",
       "      <td>2022-11-29T11:37:56.510998</td>\n",
       "      <td>pig-gut-v1-0</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>root:Host-associated:Mammals:Gastrointestinal ...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG000000114</td>\n",
       "      <td>8422</td>\n",
       "      <td>[North America]</td>\n",
       "      <td>North America</td>\n",
       "      <td>MGYG000000114</td>\n",
       "      <td>ERS2046560</td>\n",
       "      <td>ERP105624</td>\n",
       "      <td>2012808</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2021-12-07T19:57:58.564392</td>\n",
       "      <td>2021-12-07T19:57:58.561745</td>\n",
       "      <td>human-gut-v2-0-1</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>root:Host-associated:Human:Digestive system:La...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG000000113</td>\n",
       "      <td>5988</td>\n",
       "      <td>[North America, Europe, Asia, Oceania, South A...</td>\n",
       "      <td>North America</td>\n",
       "      <td>MGYG000000113</td>\n",
       "      <td>ERS2046559</td>\n",
       "      <td>ERP105624</td>\n",
       "      <td>2251845</td>\n",
       "      <td>28</td>\n",
       "      <td>...</td>\n",
       "      <td>7430.0</td>\n",
       "      <td>2021-12-07T18:48:36.632852</td>\n",
       "      <td>2021-12-07T18:48:36.630254</td>\n",
       "      <td>human-gut-v2-0-1</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>root:Host-associated:Human:Digestive system:La...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG000000109</td>\n",
       "      <td>6823</td>\n",
       "      <td>[North America]</td>\n",
       "      <td>North America</td>\n",
       "      <td>MGYG000000109</td>\n",
       "      <td>ERS2046553</td>\n",
       "      <td>ERP105624</td>\n",
       "      <td>2017268</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>338.0</td>\n",
       "      <td>2021-12-07T19:13:51.985363</td>\n",
       "      <td>2021-12-07T19:13:51.982899</td>\n",
       "      <td>human-gut-v2-0-1</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>root:Host-associated:Human:Digestive system:La...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG000000088</td>\n",
       "      <td>6812</td>\n",
       "      <td>[North America, Europe, Asia, Oceania, South A...</td>\n",
       "      <td>Europe</td>\n",
       "      <td>MGYG000000088</td>\n",
       "      <td>ERS2046340</td>\n",
       "      <td>ERP105624</td>\n",
       "      <td>1947463</td>\n",
       "      <td>51</td>\n",
       "      <td>...</td>\n",
       "      <td>2766.0</td>\n",
       "      <td>2021-12-07T19:13:34.221417</td>\n",
       "      <td>2021-12-07T19:13:34.218137</td>\n",
       "      <td>human-gut-v2-0-1</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>root:Host-associated:Human:Digestive system:La...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>genomes</td>\n",
       "      <td>MGYG000000023</td>\n",
       "      <td>6484</td>\n",
       "      <td>[North America, Europe, Asia]</td>\n",
       "      <td>Europe</td>\n",
       "      <td>MGYG000000023</td>\n",
       "      <td>ERS417304</td>\n",
       "      <td>ERP105624</td>\n",
       "      <td>2130728</td>\n",
       "      <td>26</td>\n",
       "      <td>...</td>\n",
       "      <td>2095.0</td>\n",
       "      <td>2021-12-07T19:04:28.396973</td>\n",
       "      <td>2021-12-07T19:04:28.392543</td>\n",
       "      <td>human-gut-v2-0-1</td>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>root:Host-associated:Human:Digestive system:La...</td>\n",
       "      <td>biomes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>105 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        type             id  attributes.genome-id  \\\n",
       "0    genomes  MGYG000301840                 14521   \n",
       "1    genomes  MGYG000301652                 15565   \n",
       "2    genomes  MGYG000301379                 15032   \n",
       "3    genomes  MGYG000301371                 14811   \n",
       "4    genomes  MGYG000301270                 15546   \n",
       "..       ...            ...                   ...   \n",
       "100  genomes  MGYG000000114                  8422   \n",
       "101  genomes  MGYG000000113                  5988   \n",
       "102  genomes  MGYG000000109                  6823   \n",
       "103  genomes  MGYG000000088                  6812   \n",
       "104  genomes  MGYG000000023                  6484   \n",
       "\n",
       "                           attributes.geographic-range  \\\n",
       "0                                               [Asia]   \n",
       "1                                               [Asia]   \n",
       "2                                             [Europe]   \n",
       "3                                       [Europe, Asia]   \n",
       "4                                                   []   \n",
       "..                                                 ...   \n",
       "100                                    [North America]   \n",
       "101  [North America, Europe, Asia, Oceania, South A...   \n",
       "102                                    [North America]   \n",
       "103  [North America, Europe, Asia, Oceania, South A...   \n",
       "104                      [North America, Europe, Asia]   \n",
       "\n",
       "    attributes.geographic-origin attributes.accession  \\\n",
       "0                           Asia        MGYG000301840   \n",
       "1                           Asia        MGYG000301652   \n",
       "2                         Europe        MGYG000301379   \n",
       "3                         Europe        MGYG000301371   \n",
       "4                           Asia        MGYG000301270   \n",
       "..                           ...                  ...   \n",
       "100                North America        MGYG000000114   \n",
       "101                North America        MGYG000000113   \n",
       "102                North America        MGYG000000109   \n",
       "103                       Europe        MGYG000000088   \n",
       "104                       Europe        MGYG000000023   \n",
       "\n",
       "    attributes.ena-sample-accession attributes.ena-study-accession  \\\n",
       "0                       ERS12510205                      ERP138410   \n",
       "1                       ERS12501501                      ERP137783   \n",
       "2                       ERS12500223                      ERP137406   \n",
       "3                       ERS12500425                      ERP137406   \n",
       "4                       ERS12500731                      ERP137371   \n",
       "..                              ...                            ...   \n",
       "100                      ERS2046560                      ERP105624   \n",
       "101                      ERS2046559                      ERP105624   \n",
       "102                      ERS2046553                      ERP105624   \n",
       "103                      ERS2046340                      ERP105624   \n",
       "104                       ERS417304                      ERP105624   \n",
       "\n",
       "     attributes.length  attributes.num-contigs  ...  \\\n",
       "0              2399126                      16  ...   \n",
       "1              1861661                      46  ...   \n",
       "2              1772300                     209  ...   \n",
       "3              1211138                     255  ...   \n",
       "4              1887219                      28  ...   \n",
       "..                 ...                     ...  ...   \n",
       "100            2012808                       8  ...   \n",
       "101            2251845                      28  ...   \n",
       "102            2017268                       9  ...   \n",
       "103            1947463                      51  ...   \n",
       "104            2130728                      26  ...   \n",
       "\n",
       "     attributes.pangenome-accessory-size      attributes.last-update  \\\n",
       "0                                 1072.0  2022-11-29T11:09:03.681599   \n",
       "1                                  206.0  2022-11-29T11:38:29.625963   \n",
       "2                                    0.0  2022-11-29T11:23:07.346774   \n",
       "3                                  627.0  2022-11-29T11:17:01.363252   \n",
       "4                                    NaN  2022-11-29T11:37:56.513553   \n",
       "..                                   ...                         ...   \n",
       "100                                  1.0  2021-12-07T19:57:58.564392   \n",
       "101                               7430.0  2021-12-07T18:48:36.632852   \n",
       "102                                338.0  2021-12-07T19:13:51.985363   \n",
       "103                               2766.0  2021-12-07T19:13:34.221417   \n",
       "104                               2095.0  2021-12-07T19:04:28.396973   \n",
       "\n",
       "       attributes.first-created  relationships.catalogue.data.id  \\\n",
       "0    2022-11-29T11:09:03.678953                     pig-gut-v1-0   \n",
       "1    2022-11-29T11:38:29.623194                     pig-gut-v1-0   \n",
       "2    2022-11-29T11:23:07.342356                     pig-gut-v1-0   \n",
       "3    2022-11-29T11:17:01.360465                     pig-gut-v1-0   \n",
       "4    2022-11-29T11:37:56.510998                     pig-gut-v1-0   \n",
       "..                          ...                              ...   \n",
       "100  2021-12-07T19:57:58.561745                 human-gut-v2-0-1   \n",
       "101  2021-12-07T18:48:36.630254                 human-gut-v2-0-1   \n",
       "102  2021-12-07T19:13:51.982899                 human-gut-v2-0-1   \n",
       "103  2021-12-07T19:13:34.218137                 human-gut-v2-0-1   \n",
       "104  2021-12-07T19:04:28.392543                 human-gut-v2-0-1   \n",
       "\n",
       "     relationships.catalogue.data.type  \\\n",
       "0                    genome-catalogues   \n",
       "1                    genome-catalogues   \n",
       "2                    genome-catalogues   \n",
       "3                    genome-catalogues   \n",
       "4                    genome-catalogues   \n",
       "..                                 ...   \n",
       "100                  genome-catalogues   \n",
       "101                  genome-catalogues   \n",
       "102                  genome-catalogues   \n",
       "103                  genome-catalogues   \n",
       "104                  genome-catalogues   \n",
       "\n",
       "                           relationships.biome.data.id  \\\n",
       "0    root:Host-associated:Mammals:Gastrointestinal ...   \n",
       "1    root:Host-associated:Mammals:Gastrointestinal ...   \n",
       "2    root:Host-associated:Mammals:Gastrointestinal ...   \n",
       "3    root:Host-associated:Mammals:Gastrointestinal ...   \n",
       "4    root:Host-associated:Mammals:Gastrointestinal ...   \n",
       "..                                                 ...   \n",
       "100  root:Host-associated:Human:Digestive system:La...   \n",
       "101  root:Host-associated:Human:Digestive system:La...   \n",
       "102  root:Host-associated:Human:Digestive system:La...   \n",
       "103  root:Host-associated:Human:Digestive system:La...   \n",
       "104  root:Host-associated:Human:Digestive system:La...   \n",
       "\n",
       "     relationships.biome.data.type  attributes.ncbi-genome-accession  \\\n",
       "0                           biomes                               NaN   \n",
       "1                           biomes                               NaN   \n",
       "2                           biomes                               NaN   \n",
       "3                           biomes                               NaN   \n",
       "4                           biomes                               NaN   \n",
       "..                             ...                               ...   \n",
       "100                         biomes                               NaN   \n",
       "101                         biomes                               NaN   \n",
       "102                         biomes                               NaN   \n",
       "103                         biomes                               NaN   \n",
       "104                         biomes                               NaN   \n",
       "\n",
       "     attributes.ncbi-sample-accession  attributes.ncbi-study-accession  \n",
       "0                                 NaN                              NaN  \n",
       "1                                 NaN                              NaN  \n",
       "2                                 NaN                              NaN  \n",
       "3                                 NaN                              NaN  \n",
       "4                                 NaN                              NaN  \n",
       "..                                ...                              ...  \n",
       "100                               NaN                              NaN  \n",
       "101                               NaN                              NaN  \n",
       "102                               NaN                              NaN  \n",
       "103                               NaN                              NaN  \n",
       "104                               NaN                              NaN  \n",
       "\n",
       "[105 rows x 37 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "endpoint_name = \"genomes\"\n",
    "\n",
    "with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    lineage_filter = Modifier(\"taxon_lineage=d__Bacteria;p__Firmicutes;c__Bacilli;o__Lactobacillales;f__Streptococcaceae;g__Streptococcus\")\n",
    "    resources = map(lambda r: r.json, mgnify.iterate(endpoint_name, filter=lineage_filter))\n",
    "    resources = pd.json_normalize(resources)\n",
    "resources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e34daaee-40c8-4fe7-941f-829a16a8d6fd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: >"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAigAAAGdCAYAAAA44ojeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAA00UlEQVR4nO3deXQUZb7/8U/WTiek2UmIBIhEFgVEUEMMTlgSGBUPXMBtvHPQ4ag3BBACjsIoiKMyoiwXBmTQGZQ7ol4BYWQbMmAgIGAEcURlNYgKSQSEmASSJnl+f/hLX1vCJh36SfJ+neOReurpp7/V1VX9SXVVV4AxxggAAMAigf4uAAAA4OcIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA6wT7u4BfoqKiQocPH1ZkZKQCAgL8XQ4AALgIxhj98MMPiomJUWDg+Y+R1MiAcvjwYcXGxvq7DAAA8At8/fXXatGixXn71MiAEhkZKenHBXS5XH6uBtXN7XZr7dq16tu3r0JCQvxdDgAfYvuuWwoLCxUbG+v5HD+fGhlQKr/WcblcBJQ6wO12Kzw8XC6Xix0YUMuwfddNF3N6BifJAgAA6xBQAACAdQgoAADAOgQUAABgHQIKAACwDgEFAABYh4ACAACsQ0ABAADWIaAAAADrXHJA2bhxo+68807FxMQoICBAy5Yt85pvjNHEiRPVvHlzOZ1OpaSkaN++fV59jh8/rvvvv18ul0sNGjTQsGHDVFRUdFkLAgAAao9LDijFxcW6/vrrNWfOnCrnT506VbNmzdK8efO0bds2RUREqF+/fjp9+rSnz/3336/PPvtMmZmZWrFihTZu3KiHH374ly8FAACoVS75Xjy33XabbrvttirnGWM0c+ZMPfnkkxowYIAkaeHChYqKitKyZct077336osvvtCaNWuUk5OjG2+8UZI0e/Zs3X777XrppZcUExNzGYsDAABqA5/eLDA3N1d5eXlKSUnxtNWvX18JCQnasmWL7r33Xm3ZskUNGjTwhBNJSklJUWBgoLZt26b/+I//OGvc0tJSlZaWeqYLCwsl/XiTKbfb7ctFwBVWUlKiPXv2nLdP0alSffDpAUU22Kp6Tsd5+7Zr107h4eG+LBFANarch7MvrxsuZT37NKDk5eVJkqKiorzao6KiPPPy8vLUrFkz7yKCg9WoUSNPn5+bMmWKJk+efFb72rVr+TCq4Q4cOKCxY8deVN+pF9Fn2rRpatOmzeUVBeCKy8zM9HcJuAJKSkouuq9PA0p1GT9+vDIyMjzThYWFio2NVd++feVyufxYGS5XSUmJevTocd4+e4+c1GPvfq4X/+NatW1e/7x9OYIC1Cxut1uZmZlKTU1VSEiIv8tBNav8BuRi+DSgREdHS5Ly8/PVvHlzT3t+fr66dOni6VNQUOD1uDNnzuj48eOex/+cw+GQw3H2of2QkBDe0BbLPVqs4tIzF+wXflW78853Bp6UI7pMzqvaKfwCAeXrIklF50/oEY5gxTWJuGBdAK4c9ud1w6WsY58GlLi4OEVHR2vdunWeQFJYWKht27YpLS1NkpSYmKgTJ05o+/bt6tatmyRp/fr1qqioUEJCgi/LgR/lHi1Wr5eyfDrm2MWf+mys98f1JKQAgMUuOaAUFRVp//79nunc3Fzt3LlTjRo1UsuWLTV69Gg9++yzuuaaaxQXF6ennnpKMTExGjhwoCSpQ4cO+vWvf62HHnpI8+bNk9vt1ogRI3TvvfdyBU8tUnnkZOY9XRTfrN7ljXWqVCuytqh/z0RFXOAk2QvZX1Ck0W/vvKgjOwAA/7nkgPLRRx+pV69enunKc0OGDh2q1157Tb///e9VXFyshx9+WCdOnFCPHj20Zs0ahYWFeR7zxhtvaMSIEerTp48CAwM1ePBgzZo1yweLA9vEN6unjled/2uZC3G73cprKnVt1ZBDwABQR1xyQOnZs6eMMeecHxAQoGeeeUbPPPPMOfs0atRIixYtutSnBgAAdQT34gEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWKdG3CwQNU9p+WkFhn2r3MI9Cgy7vF+SPXPmjA6fOawvjn+h4ODLe8vmFhYpMOxblZaflnR5PyAHAKg+BBRUi8PFXykibrYmfOi7MeeumeuTcSLipMPFXdRNUT4ZDwDgewQUVIuYiFYqzh2p/76ni9pc5r14zpw5o82bNiupR9JlH0E5UFCkR9/eqZherS5rHABA9SKgoFo4gsJUcfoqxbna6drGl38vntzgXHVo1OGy78VTcfqkKk5/J0dQ2IU7AwD8hpNkAQCAdQgoAADAOgQUAABgHQIKAACwDgEFAABYh4ACAACsQ0ABAADWIaAAAADrEFAAAIB1CCgAAMA6BBQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA6xBQAACAdQgoAADAOgQUAABgHQIKAACwTrC/C0DtdMpdLkna9e3Jyx6r+FSpPvpOiv7qe0U4HZc11v6CosuuBwBQ/QgoqBYH/n8QeGLppz4aMVj/sz/HR2NJEQ7e+gBgM/bSqBZ9r4uWJLVpVk/OkKDLGmvPkZMau/hTTRvSSe2a17/s2iIcwYprEnHZ4wAAqg8BBdWiUUSo7r25pU/GOnPmjCSpTdMIdbzq8gMKAMB+nCQLAACsQ0ABAADWIaAAAADrEFAAAIB1CCgAAMA6BBQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB2fB5Ty8nI99dRTiouLk9PpVJs2bfTHP/5RxhhPH2OMJk6cqObNm8vpdColJUX79u3zdSkAAKCG8nlAeeGFF/Tyyy/rz3/+s7744gu98MILmjp1qmbPnu3pM3XqVM2aNUvz5s3Ttm3bFBERoX79+un06dO+LgcAANRAwb4e8IMPPtCAAQN0xx13SJJat26tN998Ux9++KGkH4+ezJw5U08++aQGDBggSVq4cKGioqK0bNky3Xvvvb4uCQAA1DA+Dyi33HKL5s+fr71796pt27b65JNPtGnTJk2fPl2SlJubq7y8PKWkpHgeU79+fSUkJGjLli1VBpTS0lKVlpZ6pgsLCyVJbrdbbrfb14sAy5w5c8bzf9Y3ULtUbtNs23XDpaxnnweUJ554QoWFhWrfvr2CgoJUXl6u5557Tvfff78kKS8vT5IUFRXl9bioqCjPvJ+bMmWKJk+efFb72rVrFR4e7uMlgG2+LpKkYG3dulXf7vJ3NQCqQ2Zmpr9LwBVQUlJy0X19HlD+93//V2+88YYWLVqk6667Tjt37tTo0aMVExOjoUOH/qIxx48fr4yMDM90YWGhYmNj1bdvX7lcLl+VDkt9cui49OlH6t69u65v2cjf5QDwIbfbrczMTKWmpiokJMTf5aCaVX4DcjF8HlAee+wxPfHEE56vajp16qSvvvpKU6ZM0dChQxUdHS1Jys/PV/PmzT2Py8/PV5cuXaoc0+FwyOFwnNUeEhLCG7oOCA4O9vyf9Q3UTuzP64ZLWcc+v4qnpKREgYHewwYFBamiokKSFBcXp+joaK1bt84zv7CwUNu2bVNiYqKvywEAADWQz4+g3HnnnXruuefUsmVLXXfddfr44481ffp0/e53v5MkBQQEaPTo0Xr22Wd1zTXXKC4uTk899ZRiYmI0cOBAX5cDAABqIJ8HlNmzZ+upp57S8OHDVVBQoJiYGD3yyCOaOHGip8/vf/97FRcX6+GHH9aJEyfUo0cPrVmzRmFhYb4uBwAA1EA+DyiRkZGaOXOmZs6cec4+AQEBeuaZZ/TMM8/4+ukBAEAtwL14AACAdQgoAADAOgQUAABgHQIKAACwDgEFAABYh4ACAACsQ0ABAADWIaAAAADrEFAAAIB1CCgAAMA6BBQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA6xBQAACAdQgoAADAOgQUAABgHQIKAACwDgEFAABYh4ACAACsQ0ABAADWIaAAAADrEFAAAIB1CCgAAMA6BBQAAGCdYH8XgLqtpKREu3fvPm+fPUdOqDRvv77Y5VTFsQbn7du+fXuFh4f7sEIAgD8QUOBXu3fvVrdu3S6q729ev3Cf7du3q2vXrpdZFQDA3wgo8Kv27dtr+/bt5+1TdKpUK9/fojt6Jaqe03HB8QAANR8BBX4VHh5+wSMebrdb3x8tUOLNNyokJOQKVQYA8CdOkoXVysvLtWHDBm3cuFEbNmxQeXm5v0sCAFwBBBRYa+nSpYqPj1dqaqqmT5+u1NRUxcfHa+nSpf4uDQBQzQgosNLSpUs1ZMgQderUSdnZ2XrzzTeVnZ2tTp06aciQIYQUAKjlCCiwTnl5ucaOHav+/ftr2bJlSkhIkNPpVEJCgpYtW6b+/ftr3LhxfN0DALUYAQXWyc7O1sGDBzVhwgQFBnq/RQMDAzV+/Hjl5uYqOzvbTxUCAKobAQXWOXLkiCSpY8eOVc6vbK/sBwCofQgosE7z5s0lSbt27apyfmV7ZT8AQO1DQIF1br31VrVu3VrPP/+8KioqvOZVVFRoypQpiouL06233uqnCgEA1Y2AAusEBQVp2rRpWrFihQYOHKitW7fq1KlT2rp1qwYOHKgVK1bopZdeUlBQkL9LBQBUE35JFlYaNGiQFi9erLFjx+pXv/qVpz0uLk6LFy/WoEGD/FgdAKC6cQQFVjPGeE3//CsfAEDtRECBlSp/qK1z585eP9TWuXNnfqgNAOoAAgqs89MfaluyZIlOnz6tnJwcnT59WkuWLOGH2gCgDiCgwDqVP9R2yy23qG3btl734mnbtq0SExP5oTYAqOUIKLBO5Q+wTZgwocp78fzhD3/w6gcAqH0IKLBOs2bNJElJSUlV3osnKSnJqx8AoPbhMmNYraysTHPmzNH69eu1f/9+paenn3VlDwCg9qmWIyjffvut/vM//1ONGzeW0+lUp06d9NFHH3nmG2M0ceJENW/eXE6nUykpKdq3b191lIIaqKCgQJK0adMmhYeHa9y4cVq1apXGjRun8PBwbd682asfAKD28XlA+f7775WUlKSQkBCtXr1an3/+uaZNm6aGDRt6+kydOlWzZs3SvHnztG3bNkVERKhfv346ffq0r8tBDfTTe+wEBAR4zfvp3Y25Fw8A1F4+/4rnhRdeUGxsrBYsWOBpi4uL8/zbGKOZM2fqySef1IABAyRJCxcuVFRUlJYtW6Z7773X1yWhhklISJAkhYaG6sSJE9q8ebNWr16t2267TUlJSWrQoIHKyso8/QAAtY/PA8o//vEP9evXT3fddZc2bNigq666SsOHD9dDDz0kScrNzVVeXp5SUlI8j6lfv74SEhK0ZcuWKgNKaWmpSktLPdOFhYWSJLfbLbfb7etFgJ/NnTtX0o/r96677tLYsWN10003KSgoSHfddZdnnc+dO1ejRo3yZ6kALlPl9sy+vG64lPXs84Dy5Zdf6uWXX1ZGRoYmTJignJwcjRo1SqGhoRo6dKjy8vIkSVFRUV6Pi4qK8sz7uSlTpmjy5Mlnta9du1bh4eG+XgT42fr16yVJw4cP1zvvvKPevXt75kVFRWn48OGeE2fj4+P9VSYAH8rMzPR3CbgCSkpKLrqvzwNKRUWFbrzxRj3//POSpBtuuEG7du3SvHnzNHTo0F805vjx45WRkeGZLiwsVGxsrPr27SuXy+WTumGP/fv3a9WqVerUqZNeeuklZWVlKTMzU6mpqerZs6fn68PevXvr9ttv93O1AC6H2+32bN8hISH+LgfVrPIbkIvh84DSvHlzXXvttV5tHTp00JIlSyRJ0dHRkqT8/Hyvkxzz8/PVpUuXKsd0OBxyOBxntYeEhPCGroVGjhypJ554QpMmTdKwYcPUp08flZaWqk+fPgoICNDkyZMVHByskSNHsv6BWoL9ed1wKevY51fxJCUlac+ePV5te/fuVatWrST9eMJsdHS01q1b55lfWFiobdu2KTEx0dfloAYKDQ3VmDFjlJ+frxYtWujVV1/V8ePH9eqrr6pFixbKz8/XmDFjFBoa6u9SAQDVxOdHUMaMGaNbbrlFzz//vO6++259+OGHmj9/vubPny/px8tGR48erWeffVbXXHON4uLi9NRTTykmJkYDBw70dTmooaZOnSpJmjFjhoYPH+5pDw4O1mOPPeaZDwConQJMNfws54oVKzR+/Hjt27dPcXFxysjI8FzFI/14qfGkSZM0f/58nThxQj169NDcuXPVtm3bixq/sLBQ9evX18mTJzkHpZY7deqUMjIytHXrVnXv3l3Tp0+X0+n0d1kAfMTtdmvVqlW6/fbb+YqnDriUz+9qCSjVjYBSNyxdulRjx47VwYMHPW2tW7fWtGnTNGjQIP8VBsBnCCh1y6V8fnOzQFhp6dKlGjJkiPLz873a8/PzNWTIEC1dutRPlQEArgQCCqxTXl6utLQ0GWPUp08fZWdn680331R2drb69OkjY4zS0tJUXl7u71IBANWEgALrZGVlqaCgQD169NDy5cuVkJAgp9OphIQELV++XElJSSooKFBWVpa/SwUAVBMCCqxTGTwmT57sdXNA6cebBT799NNe/QAAtQ8BBQAAWIeAAuv07NlTkjRp0iRVVFR4zauoqPDcl6myHwCg9iGgwDo9e/ZU06ZNtWnTJg0YMEBbt27VqVOntHXrVg0YMECbNm1Ss2bNCCgAUIv5/JdkgcsVFBSkefPmafDgwVq3bp1WrFjhmVd59+qXX35ZQUFB/ioRAFDNOIICKw0aNEhLlixRs2bNvNqbNWumJUuW8ENtAFDLcQQF1ho0aJAGDBig999/X6tXr9Ztt92mXr16ceQEAOoAAgqsFhQUpOTkZBUXFys5OZlwAgB1BF/xAAAA6xBQAACAdQgosFp5ebk2bNigjRs3asOGDdx/BwDqCAIKrLV06VLFx8crNTVV06dPV2pqquLj47mTMQDUAQQUWGnp0qUaMmSIOnXq5HU3406dOmnIkCGEFACo5QgosE55ebnGjh2r/v37a9myZV53M162bJn69++vcePG8XUPANRiBBRYJzs7WwcPHtSECROqvJvx+PHjlZubq+zsbD9VCACobgQUWOfIkSOSpI4dO1Y5v7K9sh8AoPYhoMA6zZs3lyTt2rVLZWVlmjVrlubPn69Zs2aprKxMu3bt8uoHAKh9Aowxxt9FXKrCwkLVr19fJ0+elMvl8nc58LHy8nLFx8crKChIBw8e9DrXJCgoSK1bt1ZFRYX27dvHL8sCNZzb7daqVat0++23KyQkxN/loJpdyuc3P3UP6wQFBen666/X8uXLFRoaqiFDhsjpdOrUqVN69913deDAAQ0YMIBwAgC1GAEF1ikrK9PKlSsVHh6u0tJSvf322555wcHBCg8P18qVK1VWVqbQ0FA/VgoAqC4EFFhn7ty5OnPmjM6cOaP+/furb9++2rt3r9q2bau1a9dqxYoVnn6jR4/2b7EAgGpBQIF19u3bJ0lKTU3V8uXLVV5e7vmOOj09Xb/+9a+VmZnp6QcAqH24igfWCQgIkCR169atyt9BueGGG7z6AQBqHwIKrJOQkCBJ+tvf/qYzZ854zTtz5oxee+01r34AgNqHgALrxMbGSpIKCgrUokULvfrqqzp+/LheffVVtWjRQgUFBV79AAC1D+egwDq33nqrWrdu7fkdlOHDh3vmBQcHq02bNqqoqNCtt97qxyoBANWJgALrBAUFadq0aRoyZIhuv/12xcXFea7iyc3N1apVq7R48WJ+BwUAajECCqw0aNAgLV68WGPHjtXKlSslSWvXrlVcXJwWL16sQYMG+blCAEB1IqDAWoMGDdKAAQP0/vvva/Xq1brtttvUq1cvjpwAQB1AQIHVgoKClJycrOLiYiUnJxNOAKCO4CoeAABgHQIKAACwDgEFAABYh4ACAACsQ0ABAADWIaAAAADrEFAAAIB1CCgAAMA6BBQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA6xBQAACAdao9oPzpT39SQECARo8e7Wk7ffq00tPT1bhxY9WrV0+DBw9Wfn5+dZcCAABqiGoNKDk5OfrLX/6izp07e7WPGTNG7733nt555x1t2LBBhw8f1qBBg6qzFAAAUINUW0ApKirS/fffr1deeUUNGzb0tJ88eVJ//etfNX36dPXu3VvdunXTggUL9MEHH2jr1q3VVQ4AAKhBgqtr4PT0dN1xxx1KSUnRs88+62nfvn273G63UlJSPG3t27dXy5YttWXLFnXv3v2ssUpLS1VaWuqZLiwslCS53W653e7qWgRYonIds66B2oftu265lPVcLQHlrbfe0o4dO5STk3PWvLy8PIWGhqpBgwZe7VFRUcrLy6tyvClTpmjy5Mlnta9du1bh4eE+qRn2y8zM9HcJAKoJ23fdUFJSctF9fR5Qvv76az366KPKzMxUWFiYT8YcP368MjIyPNOFhYWKjY1V37595XK5fPIcsJfb7VZmZqZSU1MVEhLi73IA+BDbd91S+Q3IxfB5QNm+fbsKCgrUtWtXT1t5ebk2btyoP//5z/rnP/+psrIynThxwusoSn5+vqKjo6sc0+FwyOFwnNUeEhLCG7oOYX0DtRfbd91wKevY5wGlT58++vTTT73aHnzwQbVv316PP/64YmNjFRISonXr1mnw4MGSpD179ujQoUNKTEz0dTkAAKAG8nlAiYyMVMeOHb3aIiIi1LhxY0/7sGHDlJGRoUaNGsnlcmnkyJFKTEys8gRZAABQ91TbVTznM2PGDAUGBmrw4MEqLS1Vv379NHfuXH+UAgAALHRFAkpWVpbXdFhYmObMmaM5c+ZciacHAAA1DPfiAQAA1iGgAAAA6xBQAACAdQgoAADAOgQUAABgHQIKAACwDgEFAABYh4ACAACsQ0ABAADWIaAAAADrEFAAAIB1CCgAAMA6BBQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA6xBQAACAdQgoAADAOgQUAABgHQIKAACwDgEFAABYh4ACAACsQ0ABAADWIaAAAADrEFAAAIB1CCgAAMA6BBQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA6xBQAACAdQgoAADAOgQUAABgHQIKAACwDgEFAABYh4ACAACs4/OAMmXKFN10002KjIxUs2bNNHDgQO3Zs8erz+nTp5Wenq7GjRurXr16Gjx4sPLz831dCgAAqKF8HlA2bNig9PR0bd26VZmZmXK73erbt6+Ki4s9fcaMGaP33ntP77zzjjZs2KDDhw9r0KBBvi4FAADUUMG+HnDNmjVe06+99pqaNWum7du361e/+pVOnjypv/71r1q0aJF69+4tSVqwYIE6dOigrVu3qnv37r4uCQAA1DA+Dyg/d/LkSUlSo0aNJEnbt2+X2+1WSkqKp0/79u3VsmVLbdmypcqAUlpaqtLSUs90YWGhJMntdsvtdldn+bBA5TpmXQO1D9t33XIp67laA0pFRYVGjx6tpKQkdezYUZKUl5en0NBQNWjQwKtvVFSU8vLyqhxnypQpmjx58lnta9euVXh4uM/rhp0yMzP9XQKAasL2XTeUlJRcdN9qDSjp6enatWuXNm3adFnjjB8/XhkZGZ7pwsJCxcbGqm/fvnK5XJdbJizndruVmZmp1NRUhYSE+LscAD7E9l23VH4DcjGqLaCMGDFCK1as0MaNG9WiRQtPe3R0tMrKynTixAmvoyj5+fmKjo6uciyHwyGHw3FWe0hICG/oOoT1DdRebN91w6WsY59fxWOM0YgRI/Tuu+9q/fr1iouL85rfrVs3hYSEaN26dZ62PXv26NChQ0pMTPR1OQAAoAby+RGU9PR0LVq0SMuXL1dkZKTnvJL69evL6XSqfv36GjZsmDIyMtSoUSO5XC6NHDlSiYmJXMEDAAAkVUNAefnllyVJPXv29GpfsGCBHnjgAUnSjBkzFBgYqMGDB6u0tFT9+vXT3LlzfV0KAACooXweUIwxF+wTFhamOXPmaM6cOb5+egAAUAtwLx4AAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA6xBQAACAdQgoAADAOgQUAABgHQIKAACwDgEFAABYh4ACAACsQ0ABAADWIaAAAADrEFAAAIB1CCgAAMA6BBQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA6xBQAACAdQgoAADAOgQUAIBflJeXa8OGDdq4caM2bNig8vJyf5cEixBQAABX3NKlSxUfH6/U1FRNnz5dqampio+P19KlS/1dGixBQAEAXFFLly7VkCFDlJ+f79Wen5+vIUOGEFIgiYACALiCysvLlZaWJmOM+vTpo+zsbL355pvKzs5Wnz59ZIxRWloaX/eAgAIAuHKysrJUUFCgHj16aPny5UpISJDT6VRCQoKWL1+upKQkFRQUKCsry9+lws8IKACAK6YyeEyePFmBgd4fQYGBgXr66ae9+qHuIqAAAADrEFAAAFdMz549JUmTJk1SRUWF17yKigpNnjzZqx/qrmB/FwAAqJ1KSkq0e/durzaXy6WGDRtq06ZN6tmzp+69/z+195uj+jYvX2+98Xdt2rRJjRo1ksvl0o4dO7we2759e4WHh1/JRYAfEVAAANVi9+7d6tat2znnZ2dnKzs7+6z248eP6+abbz6rffv27eratatPa4S9CCgAgGrRvn17bd++vcp569ev1/Tp03XkyBFPW0xMjMaMGaPevXufczzUHQQUAEC1CA8PP+cRj65du2rMmDFa8M4KjV+0WVN+k6QH7+qvoKCgK1wlbEVAAQD4RVBQkG5M7KGIncG6MbE74QReuIoHAABYh4ACAACsw1c8AIBLlnu0WMWlZy57nAPfFXv+Hxx8+R9JEY5gxTWJuOxx4H9+DShz5szRiy++qLy8PF1//fWaPXt2lZeWAQDskXu0WL1eyvLpmGMXf+qzsd4f15OQUgv4LaC8/fbbysjI0Lx585SQkKCZM2eqX79+2rNnj5o1a+avsgAAF1B55GTmPV0U36ze5Y11qlQrsraof89ERTgdlzXW/oIijX57p0+O7MD//BZQpk+froceekgPPvigJGnevHlauXKl/va3v+mJJ57wV1kAgAsoLT+twLBvFRTWVIFhlxdQnMFnFNPwsJyReQq8zK94gsKKFBj2rUrLT0uqf1ljwf/8ElDKysq0fft2jR8/3tMWGBiolJQUbdmy5az+paWlKi0t9UwXFhZKktxut9xud/UXDL+qXMesa8AO+47vU0TcbE340Hdjzl0z1yfjRMRJXxd2Umd3I5+MB9+6lP24XwLK0aNHVV5erqioKK/2qKios+7bIElTpkzx3EDqp9auXct9GeqQzMxMf5cAQNL2fLeKD4+8YL+yo1/r2IqXfPa8jfuPU2iT2Av2Oxz6jVYdKPDZ88J3SkpKLrpvjbiKZ/z48crIyPBMFxYWKjY2Vn379pXL5fJjZbgS3G63MjMzlZqaqpCQEH+XA9R53YvL1OWLAl3dNELOkHP/uNqpUyU6eFuv845VfqZcn376qTp16qSg4PP/UFvrNtfI6Tz/H6URjiC1bswJsraq/AbkYvgloDRp0kRBQUHKz8/3as/Pz1d0dPRZ/R0OhxyOs0+eCgkJ4QOrDmF9A3aIahCi+xPjLqJnYyW2P/8RD7fbrUiV6Pbbe7J91wGXso798kNtoaGh6tatm9atW+dpq6io0Lp165SYmOiPkgAAgEX89hVPRkaGhg4dqhtvvFE333yzZs6cqeLiYs9VPQAAoO7yW0C555579N1332nixInKy8tTly5dtGbNmrNOnAUAAHWPX0+SHTFihEaMGOHPEgAAgIW4WSAAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsI5ff0n2lzLGSLq02zaj5nK73SopKVFhYSF3OwVqGbbvuqXyc7vyc/x8amRA+eGHHyRJsbHnv403AACwzw8//KD69euft0+AuZgYY5mKigodPnxYkZGRCggI8Hc5qGaFhYWKjY3V119/LZfL5e9yAPgQ23fdYozRDz/8oJiYGAUGnv8skxp5BCUwMFAtWrTwdxm4wlwuFzswoJZi+647LnTkpBInyQIAAOsQUAAAgHUIKLCew+HQpEmT5HA4/F0KAB9j+8a51MiTZAEAQO3GERQAAGAdAgoAALAOAQUAAFiHgFLDvfbaa2rQoIFn+umnn1aXLl38Vk9t0bp1a82cOdPfZaCOYru2y5XaHxw8eFABAQHauXNntT9XTUBA8aNzvRkfeOABDRw48KLGuOeee7R3716f18YH9KXhAwSV2K6r35Xe3nJycvTwww/7dMyq3g+xsbE6cuSIOnbs6NPnqqlq5C/J4kdut1tOp1NOp9PfpQDwEbZr+zRt2vSKPE9QUJCio6OvyHPVCAbVavXq1SYpKcnUr1/fNGrUyNxxxx1m//79xhhjJHn9l5ycbCZNmnRW+/vvv29yc3ONJPPWW2+ZX/3qV8bhcJgFCxaYBQsWmPr163ueb9KkSeb666838+bNMy1atDBOp9Pcdddd5sSJE54+ycnJ5tFHH/Wqc8CAAWbo0KGe+T+voVJ2drbp0aOHCQsLMy1atDAjR440RUVFnvlz5swx8fHxxuFwmGbNmpnBgwef9/XZtGmTSU5ONk6n0zRo0MD07dvXHD9+3BhjzOnTp83IkSNN06ZNjcPhMElJSebDDz/0PPb99983ksyaNWtMly5dTFhYmOnVq5fJz883q1atMu3btzeRkZHmvvvuM8XFxV7Ln56ebtLT043L5TKNGzc2Tz75pKmoqPD0adWqlZkxY4Zn+vvvvzfDhg0zTZo0MZGRkaZXr15m586dxhhjFixYcNbrtWDBggs+7qfra+HChaZVq1bG5XKZe+65xxQWFnr6lJeXm+eff960bt3ahIWFmc6dO5t33nnHM//48ePmN7/5jWnSpIkJCwsz8fHx5m9/+5sxxpjS0lKTnp5uoqOjjcPhMC1btjTPP//8edcJLoztuvq363/961+mW7duxul0msTERLN7925jzPm3t2nTppmOHTua8PBw06JFC5OWlmZ++OEHz9iVr+t7771n2rZta5xOpxk8eLApLi42r732mmnVqpVp0KCBGTlypDlz5ozncT/fH0gyr7zyihk4cKBxOp0mPj7eLF++3DP/zJkz5ne/+51nm23btq2ZOXOm1/o83/vh448/9vTNysoyN910kwkNDTXR0dHm8ccfN26322u9jxw50jz22GOmYcOGJioqykyaNOm866emIKBUs8WLF5slS5aYffv2mY8//tjceeedplOnTqa8vNx8+OGHng3xyJEj5tixY+aHH34wd999t/n1r39tjhw5Yo4cOWJKS0s9b9zWrVubJUuWmC+//NIcPny4yh1ZRESE6d27t/n444/Nhg0bTHx8vPnNb37j6XOhHdmxY8dMixYtzDPPPOOpwRhj9u/fbyIiIsyMGTPM3r17zebNm80NN9xgHnjgAWOMMTk5OSYoKMgsWrTIHDx40OzYscP893//9zlfm48//tg4HA6TlpZmdu7caXbt2mVmz55tvvvuO2OMMaNGjTIxMTFm1apV5rPPPjNDhw41DRs2NMeOHTPG/N+OrHv37mbTpk1mx44dJj4+3iQnJ5u+ffuaHTt2mI0bN5rGjRubP/3pT17LX69ePfPoo4+a3bt3m7///e8mPDzczJ8/39Pn5zuklJQUc+edd5qcnByzd+9eM3bsWNO4cWNz7NgxU1JSYsaOHWuuu+46z+tVUlJywcdVrq969eqZQYMGmU8//dRs3LjRREdHmwkTJnie+9lnnzXt27c3a9asMQcOHDALFiwwDofDZGVlGWOMSU9PN126dDE5OTkmNzfXZGZmmn/84x/GGGNefPFFExsbazZu3GgOHjxosrOzzaJFi865TnBx2K6rf7tOSEgwWVlZ5rPPPjO33nqrueWWW4wx5rzb24wZM8z69etNbm6uWbdunWnXrp1JS0vz1LZgwQITEhJiUlNTzY4dO8yGDRtM48aNTd++fc3dd99tPvvsM/Pee++Z0NBQ89Zbb3keV1VAadGihVm0aJHZt2+fGTVqlKlXr55nGcrKyszEiRNNTk6O+fLLLz37mLffftsYYy74fqgMKN98840JDw83w4cPN1988YV59913TZMmTbwCSHJysnG5XObpp582e/fuNa+//roJCAgwa9euPec6qikIKFfYd999ZySZTz/9tMq0bIwxQ4cONQMGDPBqq+z70xRujKlyRxYUFGS++eYbT9vq1atNYGCgZ4d0oR2ZMWdvkMYYM2zYMPPwww97tWVnZ5vAwEBz6tQps2TJEuNyubz++j+f++67zyQlJVU5r6ioyISEhJg33njD01ZWVmZiYmLM1KlTjTHef2lVmjJlipFkDhw44Gl75JFHTL9+/TzTycnJpkOHDl5HTB5//HHToUOHKpc/OzvbuFwuc/r0aa8a27RpY/7yl78YY/7vL9yfutjHhYeHe71mjz32mElISDDG/PjXZnh4uPnggw+8xhg2bJi57777jDHG3HnnnebBBx886zU0xpiRI0ea3r17ey0rfI/t+v9Ux3a9cuVKI8mcOnXK83r8fHuryjvvvGMaN27sma48+lJ5tMuYH/cP4eHhXkda+vXrZx555BHPdFUB5cknn/RaLklm9erV56wlPT3d68jT+d4Ple+dCRMmmHbt2nltv3PmzDH16tUz5eXlxpgf13uPHj28xrnpppvM448/fs5aagpOkq1m+/bt03333aerr75aLpdLrVu3liQdOnToF4134403XrBPy5YtddVVV3mmExMTVVFRoT179vyi56z0ySef6LXXXlO9evU8//Xr108VFRXKzc1VamqqWrVqpauvvlq//e1v9cYbb6ikpOSc4+3cuVN9+vSpct6BAwfkdruVlJTkaQsJCdHNN9+sL774wqtv586dPf+OiopSeHi4rr76aq+2goICr8d0795dAQEBnunExETt27dP5eXlVS53UVGRGjdu7LXsubm5OnDgwDmX72If17p1a0VGRnqmmzdv7ql3//79KikpUWpqqtcYCxcu9IyRlpamt956S126dNHvf/97ffDBB56xHnjgAe3cuVPt2rXTqFGjtHbt2nPWi4vHdn1lt+vmzZtL0lnb8c/961//Up8+fXTVVVcpMjJSv/3tb3Xs2DGvesPDw9WmTRvPdFRUlFq3bq169ep5tV3ouX5aX0REhFwul9dj5syZo27duqlp06aqV6+e5s+ff8nvjy+++EKJiYle+6qkpCQVFRXpm2++qbIWyXsfUpNxkmw1u/POO9WqVSu98soriomJUUVFhTp27KiysrJfNF5ERMRl1xQYGCjzszscuN3uCz6uqKhIjzzyiEaNGnXWvJYtWyo0NFQ7duxQVlaW1q5dq4kTJ+rpp59WTk6O1yWTlXx1EmBISIjn3wEBAV7TlW0VFRW/ePyioiI1b95cWVlZZ82rarku9XHnq7eoqEiStHLlSq8PJ0mee5fcdttt+uqrr7Rq1SplZmaqT58+Sk9P10svvaSuXbsqNzdXq1ev1r/+9S/dfffdSklJ0eLFiy9iyXEubNdXfruWdN7t+ODBg+rfv7/S0tL03HPPqVGjRtq0aZOGDRumsrIyhYeHnzVu5di/ZJ9xvse89dZbGjdunKZNm6bExERFRkbqxRdf1LZt2y6w1L+Mr/d5tuAISjU6duyY9uzZoyeffFJ9+vRRhw4d9P3333vmh4aGStJZf7WHhoZW+Zf8xTp06JAOHz7smd66dasCAwPVrl07ST+ekX7kyBHP/PLycu3ateuCNXTt2lWff/654uPjz/qvclmCg4OVkpKiqVOn6t///rcOHjyo9evXV1ln586dtW7duirntWnTRqGhodq8ebOnze12KycnR9dee+0lvBpV+/mOYuvWrbrmmmsUFBR0Vt+uXbsqLy9PwcHBZy13kyZNJJ379brQ4y7k2muvlcPh0KFDh84aIzY21tOvadOmGjp0qP7+979r5syZmj9/vmeey+XSPffco1deeUVvv/22lixZouPHj1/0awVvbNf+366rWo7t27eroqJC06ZNU/fu3dW2bVuv1+tK2rx5s2655RYNHz5cN9xwg+Lj48862nox74cOHTpoy5YtXsFz8+bNioyMVIsWLaqldpsQUKpRw4YN1bhxY82fP1/79+/X+vXrlZGR4ZnfrFkzOZ1OrVmzRvn5+Tp58qSkHw/5//vf/9aePXt09OjRi/or6KfCwsI0dOhQffLJJ8rOztaoUaN09913ey5f6927t1auXKmVK1dq9+7dSktL04kTJ7zGaN26tTZu3Khvv/1WR48elSQ9/vjj+uCDDzRixAjt3LlT+/bt0/LlyzVixAhJ0ooVKzRr1izt3LlTX331lRYuXKiKigrPDvTPf/6z16Hf8ePHKycnR8OHD9e///1v7d69Wy+//LKOHj2qiIgIpaWl6bHHHtOaNWv0+eef66GHHlJJSYmGDRt2aSuiCocOHVJGRob27NmjN998U7Nnz9ajjz5aZd+UlBQlJiZq4MCBWrt2rQ4ePKgPPvhAf/jDH/TRRx95Xq/c3Fzt3LlTR48eVWlp6UU97kIiIyM1btw4jRkzRq+//roOHDigHTt2aPbs2Xr99dclSRMnTtTy5cu1f/9+ffbZZ1qxYoU6dOggSZo+fbrefPNN7d69W3v37tU777yj6Ojo8x75wfmxXft/u65qe4uPj5fb7dbs2bP15Zdf6n/+5380b968S3qNfeWaa67RRx99pH/+85/au3evnnrqKeXk5Jy1DBd6PwwfPlxff/21Ro4cqd27d2v58uWaNGmSMjIyFBhYBz6+/XwOTK2XmZlpOnToYBwOh+ncubPJysoyksy7775rjDHmlVdeMbGxsSYwMNAkJycbY4wpKCgwqamppl69eue9/MyYqk+mu/76683cuXNNTEyMCQsLM0OGDPFc4mfMjyelpaWlmUaNGplmzZqZKVOmnHUy3ZYtW0znzp2Nw+Hwuhzxww8/9NQWERFhOnfubJ577jljzI8n1iUnJ5uGDRsap9NpOnfu7DlrvbK2Vq1aedWflZVlbrnlFuNwOEyDBg1Mv379zPfff2+MMebUqVNm5MiRpkmTJue9HLGyf1Wvx09fk0rJyclm+PDh5r/+67+My+UyDRs2NBMmTDjvZcaFhYVm5MiRJiYmxoSEhJjY2Fhz//33m0OHDhljfjyZdfDgwaZBgwZelz1e6HFVnew3Y8YMr9epoqLCzJw507Rr186EhISYpk2bmn79+pkNGzYYY4z54x//aDp06GCcTqdp1KiRGTBggPnyyy+NMcbMnz/fdOnSxURERBiXy2X69OljduzYYXB52K6v7Hb98ccfG0kmNzfXGHPu7W369OmmefPmxul0mn79+pmFCxd6jXUx+wdjzj6BtaqTZCvXdaX69et76jh9+rR54IEHTP369U2DBg1MWlqaeeKJJ7ye52LfDxdzmfGFTo6uqQKM+dmXlkAt17NnT3Xp0qVW/KImANRWdeAYEQAAqGkIKAAAwDp8xQMAAKzDERQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYJ3/B/MdMAKVXWMlAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "resources.boxplot(column=['attributes.completeness', 'attributes.contamination'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "adc28d42-9b70-45c4-8f5a-6c72a1d4e4bf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: >"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resources['relationships.catalogue.data.id'].hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d9a66d0-d40b-4694-aa10-efd87e4de1e0",
   "metadata": {},
   "source": [
    "## Find out whether your own MAGs are novel compared to the MGnify catalogues\n",
    "\n",
    "**The goal of this task is to query your own MAGs (created on course day 3) against MGnify's MAG catalogues, to see whether they are novel or already represented**\n",
    "\n",
    "Your MAGs generated on Day 3 should be in the directory `/home/training/Binning/contigs.fasta.metabat-bins2000.bak`. If something went wrong on Day 3 and you don't have those files any more, there are backup ones you can copy from `/media/penelopeprime/genome-resolved-metagenomics-2022/contigs.fasta.metabat-bins2000.bak`\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "A task for you\n",
    "</div>\n",
    "\n",
    "- Query the MGnify API to get a list of the IDs of all Genome Catalougues (using `jsonapi_client` as before)\n",
    "- Use the `sourmash` package to create a \"sketch\" of each of your MAGs\n",
    "- Using the `requests` library (we used this before in <a href=\"#When-the-API-endpoint-is-directly-mentioned\">the first task</a>), send your sketches to the MGnify Genome Search API with a `POST` request\n",
    "- Check the progress of your search job in a loop that waits for it to finish\n",
    "- Build a Pandas Dataframe with the search results for all of your MAGs\n",
    "\n",
    "This task involves lots of new ideas, so a lot of the code is written for you. Rather than writing the code, try and understand what each step is doing!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7462e7ef-5129-4c45-9669-0fe5a38787fe",
   "metadata": {},
   "source": [
    "### Fetch all of the catalogue IDs currently available on MGnify"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ba0ac671-6592-4fdf-afe2-8ff644a088ca",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>attributes.name</th>\n",
       "      <th>attributes.description</th>\n",
       "      <th>attributes.protein-catalogue-name</th>\n",
       "      <th>attributes.protein-catalogue-description</th>\n",
       "      <th>attributes.genome-count</th>\n",
       "      <th>attributes.unclustered-genome-count</th>\n",
       "      <th>attributes.version</th>\n",
       "      <th>attributes.last-update</th>\n",
       "      <th>attributes.pipeline-version-tag</th>\n",
       "      <th>attributes.ftp-url</th>\n",
       "      <th>relationships.biome.data.id</th>\n",
       "      <th>relationships.biome.data.type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>cow-rumen-v1-0</td>\n",
       "      <td>Cow Rumen v1.0</td>\n",
       "      <td></td>\n",
       "      <td>Cow rumen protein catalogue v1.0</td>\n",
       "      <td>Protein coding sequences from genomes in the c...</td>\n",
       "      <td>2729</td>\n",
       "      <td>5578</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2021-12-17T15:06:04</td>\n",
       "      <td>v1.2.1</td>\n",
       "      <td>http://ftp.ebi.ac.uk/pub/databases/metagenomic...</td>\n",
       "      <td>root:Host-associated:Mammals:Digestive system:...</td>\n",
       "      <td>biomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>human-gut-v2-0-1</td>\n",
       "      <td>Unified Human Gastrointestinal Genome (UHGG) v...</td>\n",
       "      <td>This is an update to the UHGG1.0 catalogue des...</td>\n",
       "      <td>UHGP v2.0.1</td>\n",
       "      <td>Protein coding sequences from genomes in the U...</td>\n",
       "      <td>4744</td>\n",
       "      <td>289232</td>\n",
       "      <td>2.0.1</td>\n",
       "      <td>2022-11-28T15:04:06</td>\n",
       "      <td>v1.2.1</td>\n",
       "      <td>http://ftp.ebi.ac.uk/pub/databases/metagenomic...</td>\n",
       "      <td>root:Host-associated:Human:Digestive system:La...</td>\n",
       "      <td>biomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>human-oral-v1-0</td>\n",
       "      <td>Human Oral v1.0</td>\n",
       "      <td></td>\n",
       "      <td>Human oral protein catalogue v1.0</td>\n",
       "      <td>Protein coding sequences from genomes in the H...</td>\n",
       "      <td>452</td>\n",
       "      <td>1225</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2021-12-17T15:07:15</td>\n",
       "      <td>v1.2.1</td>\n",
       "      <td>http://ftp.ebi.ac.uk/pub/databases/metagenomic...</td>\n",
       "      <td>root:Host-associated:Human:Digestive system:Oral</td>\n",
       "      <td>biomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>marine-v1-0</td>\n",
       "      <td>Marine v1.0</td>\n",
       "      <td></td>\n",
       "      <td>Marine protein catalogue v1.0</td>\n",
       "      <td>Protein coding sequences from genomes in the m...</td>\n",
       "      <td>1496</td>\n",
       "      <td>1504</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2021-12-17T15:06:38</td>\n",
       "      <td>v1.2.1</td>\n",
       "      <td>http://ftp.ebi.ac.uk/pub/databases/metagenomic...</td>\n",
       "      <td>root:Environmental:Aquatic:Marine</td>\n",
       "      <td>biomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>non-model-fish-gut-v1-0</td>\n",
       "      <td>Non-model Fish Gut v1.0</td>\n",
       "      <td></td>\n",
       "      <td>Non-model fish gut protein catalogue v1.0</td>\n",
       "      <td>Protein coding sequences from genomes in the n...</td>\n",
       "      <td>172</td>\n",
       "      <td>196</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2022-11-28T16:00:07</td>\n",
       "      <td>v1.3.1</td>\n",
       "      <td>http://ftp.ebi.ac.uk/pub/databases/metagenomic...</td>\n",
       "      <td>root:Host-associated:Fish:Digestive system</td>\n",
       "      <td>biomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>pig-gut-v1-0</td>\n",
       "      <td>Pig Gut v1.0</td>\n",
       "      <td></td>\n",
       "      <td>Pig gut protein catalogue v1.0</td>\n",
       "      <td>Protein coding sequences from genomes in the p...</td>\n",
       "      <td>1376</td>\n",
       "      <td>3972</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2022-11-29T09:36:43</td>\n",
       "      <td>v1.3.1</td>\n",
       "      <td>http://ftp.ebi.ac.uk/pub/databases/metagenomic...</td>\n",
       "      <td>root:Host-associated:Mammals:Gastrointestinal ...</td>\n",
       "      <td>biomes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>genome-catalogues</td>\n",
       "      <td>zebrafish-fecal-v1-0</td>\n",
       "      <td>Zebrafish Fecal v1.0</td>\n",
       "      <td></td>\n",
       "      <td>Zebrafish fecal protein catalogue v1.0</td>\n",
       "      <td>Protein coding sequences from genomes in the Z...</td>\n",
       "      <td>79</td>\n",
       "      <td>101</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2022-11-28T15:52:24</td>\n",
       "      <td>v1.3.1</td>\n",
       "      <td>http://ftp.ebi.ac.uk/pub/databases/metagenomic...</td>\n",
       "      <td>root:Host-associated:Fish:Digestive system:Int...</td>\n",
       "      <td>biomes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                type                       id  \\\n",
       "0  genome-catalogues           cow-rumen-v1-0   \n",
       "1  genome-catalogues         human-gut-v2-0-1   \n",
       "2  genome-catalogues          human-oral-v1-0   \n",
       "3  genome-catalogues              marine-v1-0   \n",
       "4  genome-catalogues  non-model-fish-gut-v1-0   \n",
       "5  genome-catalogues             pig-gut-v1-0   \n",
       "6  genome-catalogues     zebrafish-fecal-v1-0   \n",
       "\n",
       "                                     attributes.name  \\\n",
       "0                                     Cow Rumen v1.0   \n",
       "1  Unified Human Gastrointestinal Genome (UHGG) v...   \n",
       "2                                    Human Oral v1.0   \n",
       "3                                        Marine v1.0   \n",
       "4                            Non-model Fish Gut v1.0   \n",
       "5                                       Pig Gut v1.0   \n",
       "6                               Zebrafish Fecal v1.0   \n",
       "\n",
       "                              attributes.description  \\\n",
       "0                                                      \n",
       "1  This is an update to the UHGG1.0 catalogue des...   \n",
       "2                                                      \n",
       "3                                                      \n",
       "4                                                      \n",
       "5                                                      \n",
       "6                                                      \n",
       "\n",
       "           attributes.protein-catalogue-name  \\\n",
       "0           Cow rumen protein catalogue v1.0   \n",
       "1                                UHGP v2.0.1   \n",
       "2          Human oral protein catalogue v1.0   \n",
       "3              Marine protein catalogue v1.0   \n",
       "4  Non-model fish gut protein catalogue v1.0   \n",
       "5             Pig gut protein catalogue v1.0   \n",
       "6     Zebrafish fecal protein catalogue v1.0   \n",
       "\n",
       "            attributes.protein-catalogue-description  attributes.genome-count  \\\n",
       "0  Protein coding sequences from genomes in the c...                     2729   \n",
       "1  Protein coding sequences from genomes in the U...                     4744   \n",
       "2  Protein coding sequences from genomes in the H...                      452   \n",
       "3  Protein coding sequences from genomes in the m...                     1496   \n",
       "4  Protein coding sequences from genomes in the n...                      172   \n",
       "5  Protein coding sequences from genomes in the p...                     1376   \n",
       "6  Protein coding sequences from genomes in the Z...                       79   \n",
       "\n",
       "   attributes.unclustered-genome-count attributes.version  \\\n",
       "0                                 5578                1.0   \n",
       "1                               289232              2.0.1   \n",
       "2                                 1225                1.0   \n",
       "3                                 1504                1.0   \n",
       "4                                  196                1.0   \n",
       "5                                 3972                1.0   \n",
       "6                                  101                1.0   \n",
       "\n",
       "  attributes.last-update attributes.pipeline-version-tag  \\\n",
       "0    2021-12-17T15:06:04                          v1.2.1   \n",
       "1    2022-11-28T15:04:06                          v1.2.1   \n",
       "2    2021-12-17T15:07:15                          v1.2.1   \n",
       "3    2021-12-17T15:06:38                          v1.2.1   \n",
       "4    2022-11-28T16:00:07                          v1.3.1   \n",
       "5    2022-11-29T09:36:43                          v1.3.1   \n",
       "6    2022-11-28T15:52:24                          v1.3.1   \n",
       "\n",
       "                                  attributes.ftp-url  \\\n",
       "0  http://ftp.ebi.ac.uk/pub/databases/metagenomic...   \n",
       "1  http://ftp.ebi.ac.uk/pub/databases/metagenomic...   \n",
       "2  http://ftp.ebi.ac.uk/pub/databases/metagenomic...   \n",
       "3  http://ftp.ebi.ac.uk/pub/databases/metagenomic...   \n",
       "4  http://ftp.ebi.ac.uk/pub/databases/metagenomic...   \n",
       "5  http://ftp.ebi.ac.uk/pub/databases/metagenomic...   \n",
       "6  http://ftp.ebi.ac.uk/pub/databases/metagenomic...   \n",
       "\n",
       "                         relationships.biome.data.id  \\\n",
       "0  root:Host-associated:Mammals:Digestive system:...   \n",
       "1  root:Host-associated:Human:Digestive system:La...   \n",
       "2   root:Host-associated:Human:Digestive system:Oral   \n",
       "3                  root:Environmental:Aquatic:Marine   \n",
       "4         root:Host-associated:Fish:Digestive system   \n",
       "5  root:Host-associated:Mammals:Gastrointestinal ...   \n",
       "6  root:Host-associated:Fish:Digestive system:Int...   \n",
       "\n",
       "  relationships.biome.data.type  \n",
       "0                        biomes  \n",
       "1                        biomes  \n",
       "2                        biomes  \n",
       "3                        biomes  \n",
       "4                        biomes  \n",
       "5                        biomes  \n",
       "6                        biomes  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "endpoint_name = \"genome-catalogues\"\n",
    "\n",
    "with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "    resources = map(lambda r: r.json, mgnify.iterate(endpoint_name))\n",
    "    resources = pd.json_normalize(resources)\n",
    "resources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f103621c-2a00-461e-be5c-f1243bdadfd0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cow-rumen-v1-0',\n",
       " 'human-gut-v2-0-1',\n",
       " 'human-oral-v1-0',\n",
       " 'marine-v1-0',\n",
       " 'non-model-fish-gut-v1-0',\n",
       " 'pig-gut-v1-0',\n",
       " 'zebrafish-fecal-v1-0']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "catalogues = list(resources['id'])\n",
    "catalogues"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ead19559-86f3-4fef-b0be-98cafb7818b0",
   "metadata": {},
   "source": [
    "### Find all of our own MAGs, and create [\"sketches\" of them using Sourmash](https://sourmash.readthedocs.io/en/latest/index.html#sourmash-in-brief)\n",
    "We’ll compute a sourmash sketch for each MAG. A sketch goes into a signature, that we will use for searching. The signature is a sort of collection of hashes that are well suited for calculating the *containment* of your MAGs within the catalogue's MAGs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5efcefc8-df9c-432f-93fd-3b3b1d1c3abc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sourmash\n",
    "from Bio import SeqIO\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a860106e-d915-4077-b880-923b3871d888",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_mags_folder = Path('/home/training/Binning/contigs.fasta.metabat-bins2000.bak')\n",
    "\n",
    "# pathlib is a handy standard Python library for finding files and directories\n",
    "my_mags_files = list(my_mags_folder.glob('*.fa*'))\n",
    "my_mags_files\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cac471e-8b36-4ff9-8326-e10de915ec34",
   "metadata": {},
   "outputs": [],
   "source": [
    "for mag_file in my_mags_files:\n",
    "    # the sourmash parameters are chosen to match those used within MGnify\n",
    "    sketch = sourmash.MinHash(n=0, ksize=31, scaled=1000)\n",
    "    \n",
    "    # a fasta file may have multiple records in it. add them all to the sourmash signature.\n",
    "    for index, record in enumerate(SeqIO.parse(mag_file, 'fasta')):\n",
    "        sketch.add_sequence(str(record.seq))\n",
    "\n",
    "    # save the sourmash sketch as a \"signature\" file\n",
    "    sig = sourmash.SourmashSignature(sketch, name=record.name or mag_file.stem)\n",
    "    with open(mag_file.stem + '.sig', 'wt') as fp:\n",
    "        sourmash.save_signatures([sig], fp)\n",
    "\n",
    "# check what signature files we've created.\n",
    "# using ! in Jupyter lets you run a shell command. It is handy for quick things like pwd and ls.\n",
    "!ls *.sig"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69eec4eb-66ed-438c-819c-fdfd731bceed",
   "metadata": {},
   "source": [
    "### Submit a search job to the MGnify API\n",
    "We’ll call the MGnify API with all of our sketches.\n",
    "There is an endpoint for this (the same one used by the website).\n",
    "\n",
    "In this case, we need to **send** data to the API (not just fetch it). This is called \"POST\"ing data in the API world. \n",
    "\n",
    "This part of the API is quite specialized and so is not a formal JSON:API, so we use the more flexible [requests](https://docs.python-requests.org/en/master/) Python package to communicate with it, like we did earlier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e0b3f7e-2c44-409d-9804-4eb508d35713",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deb14b28-d362-4908-985a-380803ecc372",
   "metadata": {},
   "outputs": [],
   "source": [
    "endpoint = 'https://www.ebi.ac.uk/metagenomics/api/v1/genomes-search/gather'\n",
    "\n",
    "# Create a list of file uploads, and attach them to the API request\n",
    "signatures = [open(mag.stem + '.sig', 'rb') for mag in my_mags_files]\n",
    "sketch_uploads = [('file_uploaded', signature) for signature in signatures]\n",
    "\n",
    "# Send the API request - it specifies which catalogue to search against and attaches all of the signature files.\n",
    "submitted_job = requests.post(endpoint, data={'mag_catalogues': catalogues}, files=sketch_uploads).json()\n",
    "\n",
    "map(lambda fp: fp.close(), signatures)  # tidy up open file pointers\n",
    "\n",
    "print(submitted_job)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52e681b1-8fef-4deb-b349-0ca0d23f7646",
   "metadata": {},
   "source": [
    "### Wait for our results to be ready\n",
    "\n",
    "As you can see in the printed `submitted_job` above, a `status_URL` was returned in the response from submitting the job via the API.\n",
    "Since the job will be in a queue, we must poll this `status_URL` to wait for our job to be completed.\n",
    "We’ll check every 2 seconds until ALL of the jobs are finished."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecf13cb7-f5d2-49ba-9a3c-6b5c5027f423",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_done = False\n",
    "while not job_done:\n",
    "    print('Checking status...')\n",
    "    status = requests.get(submitted_job['data']['status_URL'])\n",
    "    # the status_URL is just another API endpoint that's unique for our search job\n",
    "    \n",
    "    queries_done = {sig['job_id']: sig['status'] for sig in status.json()['data']['signatures']}\n",
    "    job_done = all(map(lambda q: q == 'SUCCESS', queries_done.values()))\n",
    "    if not job_done:\n",
    "        print('Still waiting for jobs to complete. Current status of jobs')\n",
    "        print(queries_done)\n",
    "        print('Will check again in 2 seconds')\n",
    "        time.sleep(2)\n",
    "\n",
    "print('All finished!')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45cb7b2e-1e44-45fa-a677-fa2a5d74b33a",
   "metadata": {},
   "source": [
    "You can inspect the JSON data we got back by typing `status.json()` into a code cell... But essentially we have a result for each MAG against each catalogue, possibly including a best match MAG (if there was one)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9625373-ca55-4c40-849c-12437a64eb2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "search_results = pd.json_normalize(status.json()['data']['signatures'])\n",
    "search_results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfecaab1-aea2-41fd-b14e-b5aea825fdbd",
   "metadata": {},
   "source": [
    "### Are any of our MAGs found in biomes other than the human gut?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c736ae13-866e-43f7-969f-1ce7ab50cbf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "matches = search_results.dropna(subset=['result.match'])\n",
    "matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b424059-019a-4fdb-8e50-323151f7147a",
   "metadata": {},
   "outputs": [],
   "source": [
    "matches.catalogue.hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28c4dc33-6456-40a0-9126-254f1b99c5ee",
   "metadata": {},
   "source": [
    "### What is the taxonomy of the MGnify MAGs which match our query MAGs?\n",
    "To find this, we can call the API for each Genome, to find it's taxonomic lineage... We can put these results into a new column of the table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7da6fbb0-064d-47d8-80fb-4ee2981162ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_taxonomy_of_mgnify_mag(match_row):\n",
    "    mgyg_accession = match_row['result.match']\n",
    "    with APISession(\"https://www.ebi.ac.uk/metagenomics/api/v1\") as mgnify:\n",
    "        genome_document = mgnify.get('genomes', mgyg_accession)\n",
    "        return genome_document.resource.taxon_lineage\n",
    "    \n",
    "matches['best_match_taxonomy'] = matches.apply(get_taxonomy_of_mgnify_mag, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14789d3c-e464-49d2-ae8d-70ffd8dc872a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for _, match in matches.iterrows():\n",
    "    print(f\"My MAG {match['filename']} matches {match['result.match']} which has taxonomy {match['best_match_taxonomy']}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8af8846-2fd7-4684-ac55-9bc54ce22eaa",
   "metadata": {},
   "source": [
    "### Which of our MAGs are completely novel (i.e. in no MGnify catalogue)?\n",
    "\n",
    "One way to check this is group all of the search results by filename (i.e. finding the queries for each MAG vs all catalogues) and checking whether the sum of all matches is 0..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1a976c1-21bb-492a-9bfc-daf4b6a6ed63",
   "metadata": {},
   "outputs": [],
   "source": [
    "search_results.groupby('filename').apply(lambda query: query['result.matches'].sum() == 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10e8913a-bb94-4aa6-9235-e029fb1178a8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "c2e909ea68d00cf590af4aa2de30821786e45b8e11ea607135c2643544907b22"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
